{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "name": "Genomics_Workshop.ipynb",
      "provenance": [],
      "machine_shape": "hm",
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.10"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PoODL-CES/Genomics_learning_workshop/blob/main/Genomics_Workshop.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G4yBrceuFbf3"
      },
      "source": [
        "<video src=\"https://raw.githubusercontent.com/PoODL-CES/PoODL_NGS_workshop.io/main/Nature%20of%20Science.mp4\" controls autoplay loop width=\"400\" align=\"right\"></video>\n",
        "\n",
        "## Genomics Learning Workshop: NGS Analysis Pipeline\n",
        "\n",
        "Easy-to-follow workflow for processing Illumina whole-genome resequencing reads for population genomics. Steps include trimming, mapping, sorting, variant calling, variant filtering, PCA, and ADMIXTURE analysis. For more details, see the bottom of this notebook and visit the GitHub repository.\n",
        "Repository link: [Genomics Learning Workshop](https://github.com/PoODL-CES/Genomics_learning_workshop)\n",
        "\n",
        "This workshop introduces:\n",
        "- Handling raw FASTQ files\n",
        "- Performing quality control and trimming\n",
        "- Mapping reads to a reference genome\n",
        "- Calling and filtering variants\n",
        "- Running PCA and ADMIXTURE for population analysis\n"
      ]
    }
  ]
}

{
  "cell_type": "code",
  "source": [
    "#@title Install dependencies for NGS pipeline\n",
    "%%time\n",
    "import os\n",
    "\n",
    "# Variables to control optional tools (set earlier in the notebook)\n",
    "USE_BCFTOOLS = use_bcftools\n",
    "USE_PLINK = use_plink\n",
    "USE_VCFTOOLS = use_vcftools\n",
    "USE_GATK = use_gatk\n",
    "PYTHON_VERSION = python_version\n",
    "\n",
    "if not os.path.isfile(\"NGS_READY\"):\n",
    "  print(\"installing core NGS Python packages...\")\n",
    "  os.system(\"pip install -q --no-warn-conflicts pysam pandas scikit-allel matplotlib seaborn\")\n",
    "  os.system(\"touch NGS_READY\")\n",
    "\n",
    "if not os.path.isfile(\"CONDA_READY\"):\n",
    "  print(\"installing conda...\")\n",
    "  os.system(\"wget -qnc https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh\")\n",
    "  os.system(\"bash Mambaforge-Linux-x86_64.sh -bfp /usr/local\")\n",
    "  os.system(\"mamba config --set auto_update_conda false\")\n",
    "  os.system(\"touch CONDA_READY\")\n",
    "\n",
    "# Install NGS command-line tools\n",
    "if not os.path.isfile(\"FASTQC_READY\") or not os.path.isfile(\"TRIMGALORE_READY\") or not os.path.isfile(\"BWA_READY\") or not os.path.isfile(\"SAMTOOLS_READY\"):\n",
    "  print(\"installing FastQC, Trim Galore, BWA, Samtools...\")\n",
    "  os.system(f\"mamba install -y -c bioconda fastqc trim-galore bwa samtools python='{PYTHON_VERSION}'\")\n",
    "  os.system(\"touch FASTQC_READY\")\n",
    "  os.system(\"touch TRIMGALORE_READY\")\n",
    "  os.system(\"touch BWA_READY\")\n",
    "  os.system(\"touch SAMTOOLS_READY\")\n",
    "\n",
    "# Install optional variant calling and population genetics tools\n",
    "if USE_BCFTOOLS and not os.path.isfile(\"BCFTOOLS_READY\"):\n",
    "  print(\"installing bcftools...\")\n",
    "  os.system(f\"mamba install -y -c bioconda bcftools python='{PYTHON_VERSION}'\")\n",
    "  os.system(\"touch BCFTOOLS_READY\")\n",
    "\n",
    "if USE_VCFTOOLS and not os.path.isfile(\"VCFTOOLS_READY\"):\n",
    "  print(\"installing vcftools...\")\n",
    "  os.system(f\"mamba install -y -c bioconda vcftools python='{PYTHON_VERSION}'\")\n",
    "  os.system(\"touch VCFTOOLS_READY\")\n",
    "\n",
    "if USE_PLINK and not os.path.isfile(\"PLINK_READY\"):\n",
    "  print(\"installing plink...\")\n",
    "  os.system(f\"mamba install -y -c bioconda plink python='{PYTHON_VERSION}'\")\n",
    "  os.system(\"touch PLINK_READY\")\n",
    "\n",
    "if USE_GATK and not os.path.isfile(\"GATK_READY\"):\n",
    "  print(\"installing GATK...\")\n",
    "  os.system(f\"mamba install -y -c bioconda gatk4 python='{PYTHON_VERSION}'\")\n",
    "  os.system(\"touch GATK_READY\")"
  ],
  "metadata": {
    "cellView": "form",
    "id": "NGSInstallDepsFull"
  },
  "execution_count": null,
  "outputs": []
}


