{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Aksinhaa/Pony/blob/main/NGS_colab_part3.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Introduction to Variant Filtering\n",
        "\n",
        "High-throughput sequencing technologies generate millions of genetic variants across genomes, but not all detected variants are biologically meaningful or technically reliable. Raw variant calls often contain sequencing errors, low-confidence genotypes, and missing data that can severely bias downstream analyses. From removing errors to retain only high-confidence SNPs, each step here helps you build a more accurate dataset for further applications such as population genetics.\n",
        "Below are the details of the filters that we are going to use during this analysis to get a filtered VCF file.\n",
        "\n",
        "1: *passOnly*-  Retained only variant sites that passed all internal quality filter performed by the variant caller (Strelka in this case). During variant calling, Strelka assigns a PASS or FAIL tag to each variant based on multiple internal metrics such as base quality, mapping quality, strand bias, and read position bias.\n",
        "\n",
        "2: *biallelicOnly*- Kept only biallelic variants (i.e., sites with exactly two alleles) for downstream compatibility and clarity. Because most population genetic software assumes biallelic markers, this filter ensures mathematical simplicity, software compatibility, and interpretability.\n",
        "\n",
        "3: *rmvIndels*- Removed insertions and deletions (INDELs), keeping only SNPs (single nucleotide polymorphisms). Indels have higher sequencing and alignment error rates than SNPs. Removing indels ensures that the dataset is composed of high-confidence point mutations.\n",
        "\n",
        "4: *minMAF0Pt05*- Retained only variants with minor allele frequency (MAF) ≥ 0.05, meaning the alternate allele must be present in at least 5% of individuals. As very rare variants (MAF < 5%) are more likely to be sequencing errors, they are often poorly powered in statistical tests.\n",
        "\n",
        "5: *chr_E2*- Restricted to a specific chromosome/region of interest (e.g., chr_E2). This enables targeted analysis of specific genomic regions.\n",
        "\n",
        "6: *minDP3*- Required a minimum depth (DP) of 3 per genotype to avoid false positives due to low coverage. Because low-depth genotypes are highly error-prone and vulnerable to random sequencing noise.\n",
        "\n",
        "7: *minQ30*- Ensured that each site has a minimum site quality score of 30, reflecting high confidence in the variant.  This improves accuracy of allele frequency estimates and reliability of genome-wide scans.\n",
        "\n",
        "8: *minGQ30*- Filtered genotypes to retain only those with a Genotype Quality (GQ) ≥ 30, removing uncertain genotype calls. This helps prevent the misclassification of homozygous vs heterozygous states.\n",
        "\n",
        "9: *hwe_0.05*- Removed variants deviating significantly from Hardy-Weinberg Equilibrium (p < 0.05), which could indicate genotyping errors or population substructure.\n",
        "\n",
        "10: *imiss_0.6*- Filtered out individuals with >60% missing genotypes to maintain data quality. Highly incomplete samples can distort PCA and admixture results.\n",
        "\n",
        "11: *miss_0.6*- Removed variants missing in >60% of individuals. This ensures each retained SNP is well represented across individuals.\n",
        "\n",
        "12: *mid95percentile* - Likely indicates the removal of extreme outliers in genotype depth or quality, retaining the middle 95% of the data. This removes the Lowest 2.5% of undercovered, unreliable sites and the highest 2.5% of multi-mapped, repetitive, or duplicated regions.\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "IsZLf94VPkZ2"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Prerequisites and setup:\n",
        "To run this variant-filtering command line workflow, make sure you have Miniconda already installed, and it is working. Must have already downloaded the required input VCF file from Zenodo into your working directory.\n",
        "The setup involves creating and activating the required conda environments for vcftools, ggplot2, and base R, installing these packages via the bioconda and Cond forge channels."
      ],
      "metadata": {
        "id": "WQRRIryrPxk-"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bf840da6"
      },
      "source": [
        "First, we'll download and install Miniconda. We use `wget` to download the installer script and then execute it using `bash`."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f6704e35",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c477ca80-902d-421a-b471-4b54f8588e5a"
      },
      "source": [
        "# Miniconda installation and environment setup for Colab NGS Workshop\n",
        "\n",
        "# Download and install Miniconda (skip if already installed)\n",
        "!wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh\n",
        "!bash miniconda.sh -b -p /usr/local/miniconda\n",
        "\n",
        "import sys, os\n",
        "sys.path.append('/usr/local/miniconda/lib/python3.8/site-packages')\n",
        "os.environ['PATH'] = \"/usr/local/miniconda/bin:\" + os.environ['PATH']\n",
        "\n",
        "# Accept ToS for main and R conda channels\n",
        "!conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main\n",
        "!conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r\n",
        "\n",
        "# Install necessary bioinformatics tools into the environment\n",
        "!conda create -n vcf_filter -c bioconda vcftools\n"
      ],
      "execution_count": 44,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ERROR: File or directory already exists: '/usr/local/miniconda'\n",
            "If you want to update an existing installation, use the -u option.\n",
            "accepted Terms of Service for \u001b[4;94mhttps://repo.anaconda.com/pkgs/main\u001b[0m\n",
            "accepted Terms of Service for \u001b[4;94mhttps://repo.anaconda.com/pkgs/r\u001b[0m\n",
            "\u001b[1;33mJupyter detected\u001b[0m\u001b[1;33m...\u001b[0m\n",
            "\u001b[1;32m2\u001b[0m\u001b[1;32m channel Terms of Service accepted\u001b[0m\n",
            "WARNING: A conda environment already exists at '/usr/local/miniconda/envs/vcf_filter'\n",
            "\n",
            "Remove existing environment?\n",
            "This will remove ALL directories contained within this specified prefix directory, including any other conda environments.\n",
            "\n",
            " (y/[n])? y\n",
            "\n",
            "Channels:\n",
            " - bioconda\n",
            " - defaults\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Solving environment: / \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 25.9.1\n",
            "    latest version: 25.11.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/miniconda/envs/vcf_filter\n",
            "\n",
            "  added / updated specs:\n",
            "    - vcftools\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main \n",
            "  _openmp_mutex      pkgs/main/linux-64::_openmp_mutex-5.1-1_gnu \n",
            "  libgcc             pkgs/main/linux-64::libgcc-15.2.0-h69a1729_7 \n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-15.2.0-h166f726_7 \n",
            "  libgomp            pkgs/main/linux-64::libgomp-15.2.0-h4751f2c_7 \n",
            "  libnsl             pkgs/main/linux-64::libnsl-2.0.0-h5eee18b_0 \n",
            "  libstdcxx          pkgs/main/linux-64::libstdcxx-15.2.0-h39759b7_7 \n",
            "  libzlib            pkgs/main/linux-64::libzlib-1.3.1-hb25bd0a_0 \n",
            "  perl               pkgs/main/linux-64::perl-5.32.1-0_h5eee18b_perl5 \n",
            "  vcftools           bioconda/linux-64::vcftools-0.1.17-pl5321h077b44d_0 \n",
            "\n",
            "\n",
            "Proceed ([y]/n)? y\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "\n",
            "Preparing transaction: \\ \b\bdone\n",
            "Verifying transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\b| \b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate vcf_filter\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cc16c8e3"
      },
      "source": [
        "First, let's list all conda environments to confirm `vcf_filter` is present."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "46f34904",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f693f452-032e-4687-dc1c-ed512dcc879b"
      },
      "source": [
        "!conda env list"
      ],
      "execution_count": 45,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "# conda environments:\n",
            "#\n",
            "# *  -> active\n",
            "# + -> frozen\n",
            "base                     /usr/local/miniconda\n",
            "R_env                    /usr/local/miniconda/envs/R_env\n",
            "ggplot2                  /usr/local/miniconda/envs/ggplot2\n",
            "vcf_filter               /usr/local/miniconda/envs/vcf_filter\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ef02fc29"
      },
      "source": [
        "Next, we will list the packages installed in the `vcf_filter` environment to ensure `vcftools` is there."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "947ce0a6",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ed77b383-7de2-4890-808d-f0bbc9a5f53f"
      },
      "source": [
        "!conda list -n vcf_filter"
      ],
      "execution_count": 46,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# packages in environment at /usr/local/miniconda/envs/vcf_filter:\n",
            "#\n",
            "# Name                     Version          Build             Channel\n",
            "_libgcc_mutex              0.1              main\n",
            "_openmp_mutex              5.1              1_gnu\n",
            "libgcc                     15.2.0           h69a1729_7\n",
            "libgcc-ng                  15.2.0           h166f726_7\n",
            "libgomp                    15.2.0           h4751f2c_7\n",
            "libnsl                     2.0.0            h5eee18b_0\n",
            "libstdcxx                  15.2.0           h39759b7_7\n",
            "libzlib                    1.3.1            hb25bd0a_0\n",
            "perl                       5.32.1           0_h5eee18b_perl5\n",
            "vcftools                   0.1.17           pl5321h077b44d_0  bioconda\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "c3d176aa",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f71d38ed-225a-4227-c357-2456bb8f8cd4"
      },
      "source": [
        "# Create the directory if it doesn't exist\n",
        "!mkdir -p vcf_file\n",
        "\n",
        "# Download the VCF file into the created directory\n",
        "!wget -P vcf_file https://zenodo.org/records/15173226/files/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz"
      ],
      "execution_count": 47,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-12-11 10:18:38--  https://zenodo.org/records/15173226/files/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz\n",
            "Resolving zenodo.org (zenodo.org)... 188.185.43.153, 137.138.52.235, 188.185.48.75, ...\n",
            "Connecting to zenodo.org (zenodo.org)|188.185.43.153|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 104695078 (100M) [application/octet-stream]\n",
            "Saving to: ‘vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz.1’\n",
            "\n",
            "machali_Aligned_ran 100%[===================>]  99.84M  1.08MB/s    in 1m 43s  \n",
            "\n",
            "2025-12-11 10:20:21 (997 KB/s) - ‘vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz.1’ saved [104695078/104695078]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "765a8796"
      },
      "source": [
        "First, we'll apply base quality (`--minQ`), genotype quality (`--minGQ`), and Hardy-Weinberg equilibrium (`--hwe`) filters to the VCF file. The output file name will reflect these new filters."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8c999e78",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e2f55cdc-9acf-4654-ef14-f19af33058c5"
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# Define input and output files\n",
        "input_vcf_gz=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz\"\n",
        "output_prefix_step1=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\"\n",
        "\n",
        "# Run vcftools filters\n",
        "vcftools --gzvcf \"$input_vcf_gz\" \\\n",
        "         --minQ 30 \\\n",
        "         --minGQ 30 \\\n",
        "         --hwe 0.05 \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix_step1\""
      ],
      "execution_count": 48,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--gzvcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz\n",
            "\t--max-alleles 2\n",
            "\t--minGQ 30\n",
            "\t--hwe 0.05\n",
            "\t--minQ 30\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\n",
            "\t--recode\n",
            "\n",
            "Using zlib version: 1.3.1\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 60 out of 60 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 138293 out of a possible 186729 Sites\n",
            "Run Time = 64.00 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "22c1c9f3"
      },
      "source": [
        "Next, we will remove indels (insertions and deletions) from the filtered VCF file. The `--remove-indels` flag ensures that only SNP (Single Nucleotide Polymorphism) sites are retained. The output file name will again be updated to reflect this filter."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9e0f3796",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "da9e3b24-438f-4843-b0b6-b05f0b2bf7b7"
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# Define input and output files\n",
        "input_vcf_step1=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\"\n",
        "output_prefix_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\"\n",
        "\n",
        "# Remove INDELs\n",
        "vcftools --vcf \"$input_vcf_step1\" \\\n",
        "         --remove-indels \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix_step2\""
      ],
      "execution_count": 49,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\n",
            "\t--recode\n",
            "\t--remove-indels\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 60 out of 60 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 112530 out of a possible 138293 Sites\n",
            "Run Time = 49.00 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "40fc342c"
      },
      "source": [
        "Finally, we will apply the individual missingness filter. This command will generate an output file with a `.imiss` extension, which contains information about the fraction of missing sites for each individual."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "19b279ed",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e02b30a9-86ef-4974-ada1-cf7e269e6e9a"
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# Define input and output\n",
        "input_vcf_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\"\n",
        "output_prefix_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\"\n",
        "\n",
        "# Generate .imiss file\n",
        "vcftools --vcf \"$input_vcf_step2\" \\\n",
        "         --missing-indv \\\n",
        "         --out \"$output_prefix_step2\""
      ],
      "execution_count": 50,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\n",
            "\t--missing-indv\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 60 out of 60 Individuals\n",
            "Outputting Individual Missingness\n",
            "After filtering, kept 112530 out of a possible 112530 Sites\n",
            "Run Time = 1.00 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "d57d63f1"
      },
      "source": [
        "The command above will produce a file named `machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.imiss`. You can inspect this file to see the fraction of missing sites for each individual in the `F_MISS` column. This information is crucial for identifying individuals with a high proportion of missing data, which might be excluded from further analysis if the missingness exceeds a certain threshold."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "b19c2de6"
      },
      "source": [
        "Now, we will remove individuals that have a missing data proportion greater than 60%. This involves using `awk` to parse the `.imiss` file, identify individuals with `F_MISS` (fraction of missing data, which is the 5th column) greater than 0.6, and then passing these individual IDs to `vcftools` using the `--remove` flag."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "26579eef",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bb5b8516-64f8-480c-bcc8-eed1184f051a"
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# File paths\n",
        "input_vcf_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\"\n",
        "imiss_file=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.imiss\"\n",
        "output_prefix_step3=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6\"\n",
        "temp_remove_list=\"vcf_file/individuals_to_remove.txt\"\n",
        "\n",
        "# Extract individuals with >60% missing data\n",
        "awk 'NR>1 && $5>0.6 {print $1}' \"$imiss_file\" > \"$temp_remove_list\"\n",
        "\n",
        "# Remove those individuals from the VCF\n",
        "vcftools --vcf \"$input_vcf_step2\" \\\n",
        "         --remove \"$temp_remove_list\" \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix_step3\"\n",
        "\n",
        "# Cleanup temporary file\n",
        "rm \"$temp_remove_list\""
      ],
      "execution_count": 51,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\n",
            "\t--remove vcf_file/individuals_to_remove.txt\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Excluding individuals in 'exclude' list\n",
            "After filtering, kept 53 out of 60 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 112530 out of a possible 112530 Sites\n",
            "Run Time = 39.00 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "956f9310"
      },
      "source": [
        "After this step, a new VCF file will be generated (prefixed with `machali_Aligned...imiss_0.6`) that excludes individuals with more than 60% missing data. This helps to improve the quality of downstream analyses by ensuring that only individuals with a sufficient amount of genotyped data are included."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# Input VCF (after removing individuals >60% missing)\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\"\n",
        "\n",
        "# Loop over missingness thresholds (10% to 90%)\n",
        "for miss in {10..90..10}; do\n",
        "\n",
        "    # Convert percent to decimal\n",
        "    perc=$(echo \"scale=2; $miss / 100\" | bc)\n",
        "\n",
        "    # Output prefix for this missingness threshold\n",
        "    output_prefix=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_${miss}\"\n",
        "\n",
        "    # Run VCFtools missingness filter\n",
        "    vcftools --vcf \"$input_vcf\" \\\n",
        "             --max-missing \"$perc\" \\\n",
        "             --recode \\\n",
        "             --out \"$output_prefix\"\n",
        "    echo \"Completed missingness filter: $miss% (max-missing $perc)\"\n",
        "done"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UwkvZv2ress2",
        "outputId": "db7d8993-6670-4484-c902-734fc5e88fbf"
      },
      "execution_count": 52,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Completed missingness filter: 10% (max-missing .10)\n",
            "Completed missingness filter: 20% (max-missing .20)\n",
            "Completed missingness filter: 30% (max-missing .30)\n",
            "Completed missingness filter: 40% (max-missing .40)\n",
            "Completed missingness filter: 50% (max-missing .50)\n",
            "Completed missingness filter: 60% (max-missing .60)\n",
            "Completed missingness filter: 70% (max-missing .70)\n",
            "Completed missingness filter: 80% (max-missing .80)\n",
            "Completed missingness filter: 90% (max-missing .90)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.1\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_10\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 110654 out of a possible 112530 Sites\n",
            "Run Time = 40.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.2\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_20\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 109252 out of a possible 112530 Sites\n",
            "Run Time = 51.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.3\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_30\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 108069 out of a possible 112530 Sites\n",
            "Run Time = 41.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.4\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_40\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 106013 out of a possible 112530 Sites\n",
            "Run Time = 40.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.5\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_50\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 103569 out of a possible 112530 Sites\n",
            "Run Time = 38.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.6\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 99691 out of a possible 112530 Sites\n",
            "Run Time = 38.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.7\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_70\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 88717 out of a possible 112530 Sites\n",
            "Run Time = 33.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.8\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_80\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 64660 out of a possible 112530 Sites\n",
            "Run Time = 25.00 seconds\n",
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\n",
            "\t--max-missing 0.9\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_90\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 21455 out of a possible 112530 Sites\n",
            "Run Time = 10.00 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "AQhVawU4fcnk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Remove old counts file if it exists\n",
        "rm -f variant_counts.txt\n",
        "\n",
        "# Loop over missingness thresholds\n",
        "for miss in {10..90..10}; do\n",
        "\n",
        "    # Input VCF for this threshold\n",
        "    vcf_file=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_${miss}.recode.vcf\"\n",
        "\n",
        "    # Count variants (non-header lines)\n",
        "    count=$(grep -vc \"^#\" \"$vcf_file\")\n",
        "\n",
        "    # Save: missingness  variant_count\n",
        "    echo \"${miss} ${count}\" >> vcf_file/variant_counts.txt\n",
        "\n",
        "    echo \"Counted variants for ${miss}% missingness → ${count}\"\n",
        "\n",
        "done"
      ],
      "metadata": {
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-0L6UqAhfd0m",
        "outputId": "79abf972-6183-439e-dca2-ec1251eb3121"
      },
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Counted variants for 10% missingness → 110654\n",
            "Counted variants for 20% missingness → 109252\n",
            "Counted variants for 30% missingness → 108069\n",
            "Counted variants for 40% missingness → 106013\n",
            "Counted variants for 50% missingness → 103569\n",
            "Counted variants for 60% missingness → 99691\n",
            "Counted variants for 70% missingness → 88717\n",
            "Counted variants for 80% missingness → 64660\n",
            "Counted variants for 90% missingness → 21455\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "8d0ZwfaYhgRU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Add conda to PATH\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Create the ggplot2 environment\n",
        "conda create -y -n ggplot2 -c conda-forge r-ggplot2\n",
        "\n",
        "# Activate the environment\n",
        "conda activate ggplot2"
      ],
      "metadata": {
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "km2E_ExmhgmC",
        "outputId": "b5c59746-57f4-4755-b783-13393339c92d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Jupyter detected...\n",
            "2 channel Terms of Service accepted\n",
            "Channels:\n",
            " - conda-forge\n",
            " - defaults\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/miniconda/envs/ggplot2\n",
            "\n",
            "  added / updated specs:\n",
            "    - r-ggplot2\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge\n",
            "    _openmp_mutex-4.5          |            2_gnu          23 KB  conda-forge\n",
            "    _r-mutex-1.0.1             |      anacondar_1           3 KB  conda-forge\n",
            "    binutils_impl_linux-64-2.45|default_hfdba357_104         3.6 MB  conda-forge\n",
            "    bwidget-1.10.1             |       ha770c72_1         127 KB  conda-forge\n",
            "    bzip2-1.0.8                |       hda65f42_8         254 KB  conda-forge\n",
            "    c-ares-1.34.6              |       hb03c661_0         203 KB  conda-forge\n",
            "    ca-certificates-2025.11.12 |       hbd8a1cb_0         149 KB  conda-forge\n",
            "    cairo-1.18.4               |       h3394656_0         955 KB  conda-forge\n",
            "    curl-8.17.0                |       h4e3cde8_1         183 KB  conda-forge\n",
            "    font-ttf-dejavu-sans-mono-2.37|       hab24e00_0         388 KB  conda-forge\n",
            "    font-ttf-inconsolata-3.000 |       h77eed37_0          94 KB  conda-forge\n",
            "    font-ttf-source-code-pro-2.038|       h77eed37_0         684 KB  conda-forge\n",
            "    font-ttf-ubuntu-0.83       |       h77eed37_3         1.5 MB  conda-forge\n",
            "    fontconfig-2.15.0          |       h7e30c49_1         259 KB  conda-forge\n",
            "    fonts-conda-ecosystem-1    |                0           4 KB  conda-forge\n",
            "    fonts-conda-forge-1        |       hc364b38_1           4 KB  conda-forge\n",
            "    freetype-2.14.1            |       ha770c72_0         169 KB  conda-forge\n",
            "    fribidi-1.0.16             |       hb03c661_0          60 KB  conda-forge\n",
            "    gcc_impl_linux-64-15.2.0   |      hc5723f1_16        76.6 MB  conda-forge\n",
            "    gfortran_impl_linux-64-15.2.0|      h281d09f_16        19.1 MB  conda-forge\n",
            "    graphite2-1.3.14           |       hecca717_2          97 KB  conda-forge\n",
            "    gsl-2.7                    |       he838d99_0         3.2 MB  conda-forge\n",
            "    gxx_impl_linux-64-15.2.0   |      hda75c37_16        15.6 MB  conda-forge\n",
            "    harfbuzz-12.2.0            |       h15599e2_0         2.3 MB  conda-forge\n",
            "    icu-75.1                   |       he02047a_0        11.6 MB  conda-forge\n",
            "    kernel-headers_linux-64-5.14.0|       he073ed8_2         1.3 MB  conda-forge\n",
            "    keyutils-1.6.3             |       hb9d3cd8_0         131 KB  conda-forge\n",
            "    krb5-1.21.3                |       h659f571_0         1.3 MB  conda-forge\n",
            "    ld_impl_linux-64-2.45      |default_hbd61a6d_104         709 KB  conda-forge\n",
            "    lerc-4.0.0                 |       h0aef613_1         258 KB  conda-forge\n",
            "    libblas-3.11.0             |4_h4a7cf45_openblas          18 KB  conda-forge\n",
            "    libcblas-3.11.0            |4_h0358290_openblas          18 KB  conda-forge\n",
            "    libcurl-8.17.0             |       h4e3cde8_1         449 KB  conda-forge\n",
            "    libdeflate-1.25            |       h17f619e_0          72 KB  conda-forge\n",
            "    libedit-3.1.20250104       | pl5321h7949ede_0         132 KB  conda-forge\n",
            "    libev-4.33                 |       hd590300_2         110 KB  conda-forge\n",
            "    libexpat-2.7.3             |       hecca717_0          75 KB  conda-forge\n",
            "    libffi-3.5.2               |       h9ec8514_0          56 KB  conda-forge\n",
            "    libfreetype-2.14.1         |       ha770c72_0           7 KB  conda-forge\n",
            "    libfreetype6-2.14.1        |       h73754d4_0         378 KB  conda-forge\n",
            "    libgcc-15.2.0              |      he0feb66_16        1018 KB  conda-forge\n",
            "    libgcc-devel_linux-64-15.2.0|     hcc6f6b0_116         3.0 MB  conda-forge\n",
            "    libgcc-ng-15.2.0           |      h69a702a_16          27 KB  conda-forge\n",
            "    libgfortran-15.2.0         |      h69a702a_16          27 KB  conda-forge\n",
            "    libgfortran-ng-15.2.0      |      h69a702a_16          27 KB  conda-forge\n",
            "    libgfortran5-15.2.0        |      h68bc16d_16         2.4 MB  conda-forge\n",
            "    libglib-2.86.3             |       h6548e54_0         3.8 MB  conda-forge\n",
            "    libgomp-15.2.0             |      he0feb66_16         589 KB  conda-forge\n",
            "    libiconv-1.18              |       h3b78370_2         772 KB  conda-forge\n",
            "    libjpeg-turbo-3.1.2        |       hb03c661_0         619 KB  conda-forge\n",
            "    liblapack-3.11.0           |4_h47877c9_openblas          18 KB  conda-forge\n",
            "    liblzma-5.8.1              |       hb9d3cd8_2         110 KB  conda-forge\n",
            "    libnghttp2-1.67.0          |       had1ee68_0         651 KB  conda-forge\n",
            "    libopenblas-0.3.30         |pthreads_h94d23a6_4         5.7 MB  conda-forge\n",
            "    libpng-1.6.53              |       h421ea60_0         310 KB  conda-forge\n",
            "    libsanitizer-15.2.0        |      h90f66d4_16         7.3 MB  conda-forge\n",
            "    libssh2-1.11.1             |       hcf80075_0         298 KB  conda-forge\n",
            "    libstdcxx-15.2.0           |      h934c35e_16         5.6 MB  conda-forge\n",
            "    libstdcxx-devel_linux-64-15.2.0|     hd446a21_116        19.8 MB  conda-forge\n",
            "    libstdcxx-ng-15.2.0        |      hdf11a46_16          27 KB  conda-forge\n",
            "    libtiff-4.7.1              |       h9d88235_1         425 KB  conda-forge\n",
            "    libuuid-2.41.2             |       h5347b49_1          39 KB  conda-forge\n",
            "    libwebp-base-1.6.0         |       hd42ef1d_0         419 KB  conda-forge\n",
            "    libxcb-1.17.0              |       h8a09558_0         387 KB  conda-forge\n",
            "    libzlib-1.3.1              |       hb9d3cd8_2          60 KB  conda-forge\n",
            "    make-4.4.1                 |       hb9d3cd8_2         501 KB  conda-forge\n",
            "    ncurses-6.5                |       h2d0b736_3         871 KB  conda-forge\n",
            "    openssl-3.6.0              |       h26f9b46_0         3.0 MB  conda-forge\n",
            "    pango-1.56.4               |       hadf4263_0         445 KB  conda-forge\n",
            "    pcre2-10.47                |       haa7fec5_0         1.2 MB  conda-forge\n",
            "    pixman-0.46.4              |       h54a6638_1         440 KB  conda-forge\n",
            "    pthread-stubs-0.4          |    hb9d3cd8_1002           8 KB  conda-forge\n",
            "    r-base-4.5.2               |       h835929b_2        26.0 MB  conda-forge\n",
            "    r-cli-3.6.5                |    r45h3697838_1         1.3 MB  conda-forge\n",
            "    r-colorspace-2.1_2         |    r45h54b55ab_0         2.4 MB  conda-forge\n",
            "    r-farver-2.1.2             |    r45h3697838_2         1.4 MB  conda-forge\n",
            "    r-ggplot2-4.0.1            |    r45h785f33e_0         7.5 MB  conda-forge\n",
            "    r-glue-1.8.0               |    r45h54b55ab_1         161 KB  conda-forge\n",
            "    r-gtable-0.3.6             |    r45hc72bb7e_1         224 KB  conda-forge\n",
            "    r-isoband-0.2.7            |    r45h3697838_4         1.6 MB  conda-forge\n",
            "    r-labeling-0.4.3           |    r45hc72bb7e_2          69 KB  conda-forge\n",
            "    r-lifecycle-1.0.4          |    r45hc72bb7e_2         121 KB  conda-forge\n",
            "    r-munsell-0.5.1            |    r45hc72bb7e_2         241 KB  conda-forge\n",
            "    r-r6-2.6.1                 |    r45hc72bb7e_1          93 KB  conda-forge\n",
            "    r-rcolorbrewer-1.1_3       |    r45h785f33e_4          66 KB  conda-forge\n",
            "    r-rlang-1.1.6              |    r45h3697838_1         1.5 MB  conda-forge\n",
            "    r-s7-0.2.1                 |    r45h54b55ab_0         304 KB  conda-forge\n",
            "    r-scales-1.4.0             |    r45hc72bb7e_1         760 KB  conda-forge\n",
            "    r-vctrs-0.6.5              |    r45h3697838_2         1.2 MB  conda-forge\n",
            "    r-viridislite-0.4.2        |    r45hc72bb7e_3         1.2 MB  conda-forge\n",
            "    r-withr-3.0.2              |    r45hc72bb7e_1         230 KB  conda-forge\n",
            "    readline-8.2               |       h8c095d6_2         276 KB  conda-forge\n",
            "    sed-4.9                    |       h6688a6e_0         224 KB  conda-forge\n",
            "    sysroot_linux-64-2.34      |       h087de78_2        38.0 MB  conda-forge\n",
            "    tk-8.6.13                  |noxft_ha0e22de_103         3.1 MB  conda-forge\n",
            "    tktable-2.10               |       h8d826fa_7          90 KB  conda-forge\n",
            "    tzdata-2025b               |       h78e105d_0         120 KB  conda-forge\n",
            "    xorg-libice-1.1.2          |       hb9d3cd8_0          57 KB  conda-forge\n",
            "    xorg-libsm-1.2.6           |       he73a12e_0          27 KB  conda-forge\n",
            "    xorg-libx11-1.8.12         |       h4f16b4b_0         816 KB  conda-forge\n",
            "    xorg-libxau-1.0.12         |       hb03c661_1          15 KB  conda-forge\n",
            "    xorg-libxdmcp-1.1.5        |       hb03c661_1          20 KB  conda-forge\n",
            "    xorg-libxext-1.3.6         |       hb9d3cd8_0          49 KB  conda-forge\n",
            "    xorg-libxrender-0.9.12     |       hb9d3cd8_0          32 KB  conda-forge\n",
            "    xorg-libxt-1.3.1           |       hb9d3cd8_0         371 KB  conda-forge\n",
            "    zstd-1.5.7                 |       hb78ec9c_6         587 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:       291.5 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge \n",
            "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-2_gnu \n",
            "  _r-mutex           conda-forge/noarch::_r-mutex-1.0.1-anacondar_1 \n",
            "  binutils_impl_lin~ conda-forge/linux-64::binutils_impl_linux-64-2.45-default_hfdba357_104 \n",
            "  bwidget            conda-forge/linux-64::bwidget-1.10.1-ha770c72_1 \n",
            "  bzip2              conda-forge/linux-64::bzip2-1.0.8-hda65f42_8 \n",
            "  c-ares             conda-forge/linux-64::c-ares-1.34.6-hb03c661_0 \n",
            "  ca-certificates    conda-forge/noarch::ca-certificates-2025.11.12-hbd8a1cb_0 \n",
            "  cairo              conda-forge/linux-64::cairo-1.18.4-h3394656_0 \n",
            "  curl               conda-forge/linux-64::curl-8.17.0-h4e3cde8_1 \n",
            "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0 \n",
            "  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0 \n",
            "  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0 \n",
            "  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-h77eed37_3 \n",
            "  fontconfig         conda-forge/linux-64::fontconfig-2.15.0-h7e30c49_1 \n",
            "  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0 \n",
            "  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-hc364b38_1 \n",
            "  freetype           conda-forge/linux-64::freetype-2.14.1-ha770c72_0 \n",
            "  fribidi            conda-forge/linux-64::fribidi-1.0.16-hb03c661_0 \n",
            "  gcc_impl_linux-64  conda-forge/linux-64::gcc_impl_linux-64-15.2.0-hc5723f1_16 \n",
            "  gfortran_impl_lin~ conda-forge/linux-64::gfortran_impl_linux-64-15.2.0-h281d09f_16 \n",
            "  graphite2          conda-forge/linux-64::graphite2-1.3.14-hecca717_2 \n",
            "  gsl                conda-forge/linux-64::gsl-2.7-he838d99_0 \n",
            "  gxx_impl_linux-64  conda-forge/linux-64::gxx_impl_linux-64-15.2.0-hda75c37_16 \n",
            "  harfbuzz           conda-forge/linux-64::harfbuzz-12.2.0-h15599e2_0 \n",
            "  icu                conda-forge/linux-64::icu-75.1-he02047a_0 \n",
            "  kernel-headers_li~ conda-forge/noarch::kernel-headers_linux-64-5.14.0-he073ed8_2 \n",
            "  keyutils           conda-forge/linux-64::keyutils-1.6.3-hb9d3cd8_0 \n",
            "  krb5               conda-forge/linux-64::krb5-1.21.3-h659f571_0 \n",
            "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.45-default_hbd61a6d_104 \n",
            "  lerc               conda-forge/linux-64::lerc-4.0.0-h0aef613_1 \n",
            "  libblas            conda-forge/linux-64::libblas-3.11.0-4_h4a7cf45_openblas \n",
            "  libcblas           conda-forge/linux-64::libcblas-3.11.0-4_h0358290_openblas \n",
            "  libcurl            conda-forge/linux-64::libcurl-8.17.0-h4e3cde8_1 \n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.25-h17f619e_0 \n",
            "  libedit            conda-forge/linux-64::libedit-3.1.20250104-pl5321h7949ede_0 \n",
            "  libev              conda-forge/linux-64::libev-4.33-hd590300_2 \n",
            "  libexpat           conda-forge/linux-64::libexpat-2.7.3-hecca717_0 \n",
            "  libffi             conda-forge/linux-64::libffi-3.5.2-h9ec8514_0 \n",
            "  libfreetype        conda-forge/linux-64::libfreetype-2.14.1-ha770c72_0 \n",
            "  libfreetype6       conda-forge/linux-64::libfreetype6-2.14.1-h73754d4_0 \n",
            "  libgcc             conda-forge/linux-64::libgcc-15.2.0-he0feb66_16 \n",
            "  libgcc-devel_linu~ conda-forge/noarch::libgcc-devel_linux-64-15.2.0-hcc6f6b0_116 \n",
            "  libgcc-ng          conda-forge/linux-64::libgcc-ng-15.2.0-h69a702a_16 \n",
            "  libgfortran        conda-forge/linux-64::libgfortran-15.2.0-h69a702a_16 \n",
            "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-15.2.0-h69a702a_16 \n",
            "  libgfortran5       conda-forge/linux-64::libgfortran5-15.2.0-h68bc16d_16 \n",
            "  libglib            conda-forge/linux-64::libglib-2.86.3-h6548e54_0 \n",
            "  libgomp            conda-forge/linux-64::libgomp-15.2.0-he0feb66_16 \n",
            "  libiconv           conda-forge/linux-64::libiconv-1.18-h3b78370_2 \n",
            "  libjpeg-turbo      conda-forge/linux-64::libjpeg-turbo-3.1.2-hb03c661_0 \n",
            "  liblapack          conda-forge/linux-64::liblapack-3.11.0-4_h47877c9_openblas \n",
            "  liblzma            conda-forge/linux-64::liblzma-5.8.1-hb9d3cd8_2 \n",
            "  libnghttp2         conda-forge/linux-64::libnghttp2-1.67.0-had1ee68_0 \n",
            "  libopenblas        conda-forge/linux-64::libopenblas-0.3.30-pthreads_h94d23a6_4 \n",
            "  libpng             conda-forge/linux-64::libpng-1.6.53-h421ea60_0 \n",
            "  libsanitizer       conda-forge/linux-64::libsanitizer-15.2.0-h90f66d4_16 \n",
            "  libssh2            conda-forge/linux-64::libssh2-1.11.1-hcf80075_0 \n",
            "  libstdcxx          conda-forge/linux-64::libstdcxx-15.2.0-h934c35e_16 \n",
            "  libstdcxx-devel_l~ conda-forge/noarch::libstdcxx-devel_linux-64-15.2.0-hd446a21_116 \n",
            "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-15.2.0-hdf11a46_16 \n",
            "  libtiff            conda-forge/linux-64::libtiff-4.7.1-h9d88235_1 \n",
            "  libuuid            conda-forge/linux-64::libuuid-2.41.2-h5347b49_1 \n",
            "  libwebp-base       conda-forge/linux-64::libwebp-base-1.6.0-hd42ef1d_0 \n",
            "  libxcb             conda-forge/linux-64::libxcb-1.17.0-h8a09558_0 \n",
            "  libzlib            conda-forge/linux-64::libzlib-1.3.1-hb9d3cd8_2 \n",
            "  make               conda-forge/linux-64::make-4.4.1-hb9d3cd8_2 \n",
            "  ncurses            conda-forge/linux-64::ncurses-6.5-h2d0b736_3 \n",
            "  openssl            conda-forge/linux-64::openssl-3.6.0-h26f9b46_0 \n",
            "  pango              conda-forge/linux-64::pango-1.56.4-hadf4263_0 \n",
            "  pcre2              conda-forge/linux-64::pcre2-10.47-haa7fec5_0 \n",
            "  pixman             conda-forge/linux-64::pixman-0.46.4-h54a6638_1 \n",
            "  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-hb9d3cd8_1002 \n",
            "  r-base             conda-forge/linux-64::r-base-4.5.2-h835929b_2 \n",
            "  r-cli              conda-forge/linux-64::r-cli-3.6.5-r45h3697838_1 \n",
            "  r-colorspace       conda-forge/linux-64::r-colorspace-2.1_2-r45h54b55ab_0 \n",
            "  r-farver           conda-forge/linux-64::r-farver-2.1.2-r45h3697838_2 \n",
            "  r-ggplot2          conda-forge/noarch::r-ggplot2-4.0.1-r45h785f33e_0 \n",
            "  r-glue             conda-forge/linux-64::r-glue-1.8.0-r45h54b55ab_1 \n",
            "  r-gtable           conda-forge/noarch::r-gtable-0.3.6-r45hc72bb7e_1 \n",
            "  r-isoband          conda-forge/linux-64::r-isoband-0.2.7-r45h3697838_4 \n",
            "  r-labeling         conda-forge/noarch::r-labeling-0.4.3-r45hc72bb7e_2 \n",
            "  r-lifecycle        conda-forge/noarch::r-lifecycle-1.0.4-r45hc72bb7e_2 \n",
            "  r-munsell          conda-forge/noarch::r-munsell-0.5.1-r45hc72bb7e_2 \n",
            "  r-r6               conda-forge/noarch::r-r6-2.6.1-r45hc72bb7e_1 \n",
            "  r-rcolorbrewer     conda-forge/noarch::r-rcolorbrewer-1.1_3-r45h785f33e_4 \n",
            "  r-rlang            conda-forge/linux-64::r-rlang-1.1.6-r45h3697838_1 \n",
            "  r-s7               conda-forge/linux-64::r-s7-0.2.1-r45h54b55ab_0 \n",
            "  r-scales           conda-forge/noarch::r-scales-1.4.0-r45hc72bb7e_1 \n",
            "  r-vctrs            conda-forge/linux-64::r-vctrs-0.6.5-r45h3697838_2 \n",
            "  r-viridislite      conda-forge/noarch::r-viridislite-0.4.2-r45hc72bb7e_3 \n",
            "  r-withr            conda-forge/noarch::r-withr-3.0.2-r45hc72bb7e_1 \n",
            "  readline           conda-forge/linux-64::readline-8.2-h8c095d6_2 \n",
            "  sed                conda-forge/linux-64::sed-4.9-h6688a6e_0 \n",
            "  sysroot_linux-64   conda-forge/noarch::sysroot_linux-64-2.34-h087de78_2 \n",
            "  tk                 conda-forge/linux-64::tk-8.6.13-noxft_ha0e22de_103 \n",
            "  tktable            conda-forge/linux-64::tktable-2.10-h8d826fa_7 \n",
            "  tzdata             conda-forge/noarch::tzdata-2025b-h78e105d_0 \n",
            "  xorg-libice        conda-forge/linux-64::xorg-libice-1.1.2-hb9d3cd8_0 \n",
            "  xorg-libsm         conda-forge/linux-64::xorg-libsm-1.2.6-he73a12e_0 \n",
            "  xorg-libx11        conda-forge/linux-64::xorg-libx11-1.8.12-h4f16b4b_0 \n",
            "  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.12-hb03c661_1 \n",
            "  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.5-hb03c661_1 \n",
            "  xorg-libxext       conda-forge/linux-64::xorg-libxext-1.3.6-hb9d3cd8_0 \n",
            "  xorg-libxrender    conda-forge/linux-64::xorg-libxrender-0.9.12-hb9d3cd8_0 \n",
            "  xorg-libxt         conda-forge/linux-64::xorg-libxt-1.3.1-hb9d3cd8_0 \n",
            "  zstd               conda-forge/linux-64::zstd-1.5.7-hb78ec9c_6 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages: ...working...\rgcc_impl_linux-64-15 | 76.6 MB   |            |   0% \n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   |            |   0% \u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   |            |   0% \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   |            |   0% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\rgfortran_impl_linux- | 19.1 MB   |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-ggplot2-4.0.1      | 7.5 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibsanitizer-15.2.0  | 7.3 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibopenblas-0.3.30   | 5.7 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibstdcxx-15.2.0     | 5.6 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibglib-2.86.3       | 3.8 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rbinutils_impl_linux- | 3.6 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgsl-2.7              | 3.2 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rtk-8.6.13            | 3.1 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ropenssl-3.6.0        | 3.0 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgcc-devel_linux-6 | 3.0 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-colorspace-2.1_2   | 2.4 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgfortran5-15.2.0  | 2.4 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\rgfortran_impl_linux- | 19.1 MB   |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   |            |   0% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   |            |   0% \u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   |            |   0% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\rgfortran_impl_linux- | 19.1 MB   | ####2      |  43% \u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   |            |   0% \n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | #8         |  18% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | 8          |   8% \u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | 4          |   5% \u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | 2          |   3% \n",
            "\n",
            "\n",
            "\n",
            "\rgfortran_impl_linux- | 19.1 MB   | ######8    |  68% \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | ###3       |  33% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | ##         |  21% \u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | #4         |  15% \u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | 8          |   8% \n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | #####1     |  51% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\rgfortran_impl_linux- | 19.1 MB   | ########9  |  89% \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | ###        |  31% \u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ##3        |  23% \u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #3         |  14% \n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | ######5    |  65% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | ####4      |  45% \u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ###        |  31% \u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #9         |  19% \n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | ########1  |  82% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | #####8     |  59% \u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ###8       |  38% \u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ##3        |  24% \n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | #######1   |  71% \u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ####8      |  48% \u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ##9        |  29% \n",
            "\n",
            "\n",
            "\n",
            "\rgfortran_impl_linux- | 19.1 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | ########6  |  87% \u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ###4       |  35% \n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | #####6     |  56% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | #########9 | 100% \u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ###9       |  40% \n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ######4    |  64% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | #5         |  16% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | #######1   |  72% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | ###2       |  33% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ####4      |  44% \n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | #######8   |  79% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | ####8      |  48% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ####8      |  49% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   | ##5        |  25% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ########5  |  86% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | ######1    |  62% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #####2     |  53% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   | ####7      |  47% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | #########2 |  93% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | ########2  |  83% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #####6     |  57% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   | #######6   |  76% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | #########9 | 100% \u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ######     |  61% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   | #########7 |  98% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ######4    |  64% \n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | ########## | 100% \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-ggplot2-4.0.1      | 7.5 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ######7    |  68% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-ggplot2-4.0.1      | 7.5 MB    | ###5       |  36% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibsanitizer-15.2.0  | 7.3 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibopenblas-0.3.30   | 5.7 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #######    |  71% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-ggplot2-4.0.1      | 7.5 MB    | ######2    |  62% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibsanitizer-15.2.0  | 7.3 MB    | ##9        |  30% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibopenblas-0.3.30   | 5.7 MB    | ###2       |  32% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibsanitizer-15.2.0  | 7.3 MB    | ####8      |  49% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #######3   |  74% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibopenblas-0.3.30   | 5.7 MB    | ######8    |  69% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibsanitizer-15.2.0  | 7.3 MB    | ########3  |  84% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #######7   |  77% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-ggplot2-4.0.1      | 7.5 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-ggplot2-4.0.1      | 7.5 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ########## | 100% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibstdcxx-15.2.0     | 5.6 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #######9   |  80% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibopenblas-0.3.30   | 5.7 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibopenblas-0.3.30   | 5.7 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibglib-2.86.3       | 3.8 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rbinutils_impl_linux- | 3.6 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibsanitizer-15.2.0  | 7.3 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibstdcxx-15.2.0     | 5.6 MB    | ##9        |  30% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ########2  |  83% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgsl-2.7              | 3.2 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibglib-2.86.3       | 3.8 MB    | ########5  |  86% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rbinutils_impl_linux- | 3.6 MB    | #######1   |  72% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibstdcxx-15.2.0     | 5.6 MB    | #######4   |  75% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ########5  |  86% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgsl-2.7              | 3.2 MB    | #######9   |  80% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ########8  |  89% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibglib-2.86.3       | 3.8 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rbinutils_impl_linux- | 3.6 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgsl-2.7              | 3.2 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibstdcxx-15.2.0     | 5.6 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ropenssl-3.6.0        | 3.0 MB    |            |   1% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rtk-8.6.13            | 3.1 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #########1 |  91% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgcc-devel_linux-6 | 3.0 MB    |            |   1% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-colorspace-2.1_2   | 2.4 MB    |            |   1% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rtk-8.6.13            | 3.1 MB    | ######2    |  63% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #########4 |  94% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgcc-devel_linux-6 | 3.0 MB    | ######6    |  67% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ropenssl-3.6.0        | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ropenssl-3.6.0        | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-colorspace-2.1_2   | 2.4 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-colorspace-2.1_2   | 2.4 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgfortran5-15.2.0  | 2.4 MB    |            |   1% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #########6 |  97% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgcc-devel_linux-6 | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rtk-8.6.13            | 3.1 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rtk-8.6.13            | 3.1 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgfortran5-15.2.0  | 2.4 MB    | ########6  |  87% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | #########9 |  99% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgfortran5-15.2.0  | 2.4 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\rgfortran_impl_linux- | 19.1 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ########## | 100% \n",
            "\n",
            "\n",
            "\rlibstdcxx-devel_linu | 19.8 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\rr-base-4.5.2         | 26.0 MB   | ########## | 100% \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ricu-75.1             | 11.6 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-ggplot2-4.0.1      | 7.5 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgxx_impl_linux-64-15 | 15.6 MB   | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibopenblas-0.3.30   | 5.7 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibsanitizer-15.2.0  | 7.3 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibglib-2.86.3       | 3.8 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rbinutils_impl_linux- | 3.6 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rgsl-2.7              | 3.2 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\rsysroot_linux-64-2.3 | 38.0 MB   | ########## | 100% \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibstdcxx-15.2.0     | 5.6 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rr-colorspace-2.1_2   | 2.4 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\ropenssl-3.6.0        | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgcc-devel_linux-6 | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rlibgfortran5-15.2.0  | 2.4 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\rtk-8.6.13            | 3.1 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\rgcc_impl_linux-64-15 | 76.6 MB   | ########## | 100% \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                      \r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\r                                                     \r\n",
            "\r                                                     \u001b[A\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\r\r\n",
            "\r\u001b[A\n",
            "\n",
            "\r\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\r\r\n",
            "\r\u001b[A\n",
            "\n",
            "\r\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\r\r\n",
            "\r\u001b[A\n",
            "\n",
            "\r\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\r\r\n",
            "\r\u001b[A\n",
            "\n",
            "\r\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\r\r\n",
            "\r\u001b[A\n",
            "\n",
            "\r\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\r\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A done\n",
            "Preparing transaction: - \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Verifying transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate ggplot2\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 25.9.1\n",
            "    latest version: 25.11.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "Q6JTdlbHib6E"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda ggplot2 environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate ggplot2\n",
        "\n",
        "# Create an R script to run ggplot2\n",
        "cat << 'EOF' > plot_variants.R\n",
        "library(ggplot2)\n",
        "library(scales)\n",
        "\n",
        "data <- read.table(\"vcf_file/variant_counts.txt\", header=FALSE,\n",
        "                   col.names=c(\"Missingness\", \"Variants\"))\n",
        "\n",
        "p <- ggplot(data, aes(x=Missingness, y=Variants)) +\n",
        "     geom_line(color=\"blue\") +\n",
        "     geom_point(color=\"red\") +\n",
        "     ggtitle(\"Number of Passed Variants vs. Missingness Filter\") +\n",
        "     xlab(\"Max Missingness (%)\") +\n",
        "     ylab(\"Number of Variants\") +\n",
        "     scale_y_continuous(labels = comma) +\n",
        "     scale_x_continuous(limits = c(10, 90)) +\n",
        "     theme_minimal()\n",
        "\n",
        "ggsave(\"vcf_file/variant_plot.png\", plot = p)\n",
        "EOF\n",
        "\n",
        "# Run the R script\n",
        "Rscript plot_variants.R"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Hwyundz6icRV",
        "outputId": "c534bd1b-f8d2-4b83-e88d-c14607506198"
      },
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Saving 7 x 7 in image\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "5ekVd2gUjXrs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# Use the correct miss_60 file\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60.recode.vcf\"\n",
        "\n",
        "output_prefix=\"vcf_file/depth\"\n",
        "\n",
        "vcftools --vcf \"$input_vcf\" \\\n",
        "         --site-mean-depth \\\n",
        "         --out \"$output_prefix\"\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "9UdDgXgwjYBz",
        "outputId": "0094e0c0-0841-4227-8628-9583e98a313b"
      },
      "execution_count": 57,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60.recode.vcf\n",
            "\t--out vcf_file/depth\n",
            "\t--site-mean-depth\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting Depth for Each Site\n",
            "After filtering, kept 99691 out of a possible 99691 Sites\n",
            "Run Time = 2.00 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "dspFE9IClDej"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Add conda to PATH and initialize shell\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Create R environment (run once)\n",
        "conda create -y -n R_env -c conda-forge r-base\n",
        "\n",
        "# Activate R environment\n",
        "conda activate R_env\n",
        "\n",
        "# Create an R script to calculate middle 95% depth\n",
        "cat << 'EOF' > compute_depth_range.R\n",
        "depth_data <- read.table(\"vcf_file/depth.ldepth.mean\", header = TRUE)\n",
        "\n",
        "depths <- depth_data$MEAN_DEPTH\n",
        "\n",
        "lower_bound <- quantile(depths, 0.025, na.rm = TRUE)\n",
        "upper_bound <- quantile(depths, 0.975, na.rm = TRUE)\n",
        "\n",
        "cat(\"Middle 95% range:\", lower_bound, \"to\", upper_bound, \"\\n\")\n",
        "EOF\n",
        "\n",
        "# Run the R script\n",
        "Rscript compute_depth_range.R\n"
      ],
      "metadata": {
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UME3l06plD2d",
        "outputId": "44646cbd-d1cc-42f4-aa67-8981c872ed6e"
      },
      "execution_count": 58,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Jupyter detected...\n",
            "2 channel Terms of Service accepted\n",
            "Channels:\n",
            " - conda-forge\n",
            " - defaults\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/miniconda/envs/R_env\n",
            "\n",
            "  added / updated specs:\n",
            "    - r-base\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge \n",
            "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-2_gnu \n",
            "  _r-mutex           conda-forge/noarch::_r-mutex-1.0.1-anacondar_1 \n",
            "  binutils_impl_lin~ conda-forge/linux-64::binutils_impl_linux-64-2.45-default_hfdba357_104 \n",
            "  bwidget            conda-forge/linux-64::bwidget-1.10.1-ha770c72_1 \n",
            "  bzip2              conda-forge/linux-64::bzip2-1.0.8-hda65f42_8 \n",
            "  c-ares             conda-forge/linux-64::c-ares-1.34.6-hb03c661_0 \n",
            "  ca-certificates    conda-forge/noarch::ca-certificates-2025.11.12-hbd8a1cb_0 \n",
            "  cairo              conda-forge/linux-64::cairo-1.18.4-h3394656_0 \n",
            "  curl               conda-forge/linux-64::curl-8.17.0-h4e3cde8_1 \n",
            "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0 \n",
            "  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0 \n",
            "  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0 \n",
            "  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-h77eed37_3 \n",
            "  fontconfig         conda-forge/linux-64::fontconfig-2.15.0-h7e30c49_1 \n",
            "  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0 \n",
            "  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-hc364b38_1 \n",
            "  freetype           conda-forge/linux-64::freetype-2.14.1-ha770c72_0 \n",
            "  fribidi            conda-forge/linux-64::fribidi-1.0.16-hb03c661_0 \n",
            "  gcc_impl_linux-64  conda-forge/linux-64::gcc_impl_linux-64-15.2.0-hc5723f1_16 \n",
            "  gfortran_impl_lin~ conda-forge/linux-64::gfortran_impl_linux-64-15.2.0-h281d09f_16 \n",
            "  graphite2          conda-forge/linux-64::graphite2-1.3.14-hecca717_2 \n",
            "  gsl                conda-forge/linux-64::gsl-2.7-he838d99_0 \n",
            "  gxx_impl_linux-64  conda-forge/linux-64::gxx_impl_linux-64-15.2.0-hda75c37_16 \n",
            "  harfbuzz           conda-forge/linux-64::harfbuzz-12.2.0-h15599e2_0 \n",
            "  icu                conda-forge/linux-64::icu-75.1-he02047a_0 \n",
            "  kernel-headers_li~ conda-forge/noarch::kernel-headers_linux-64-5.14.0-he073ed8_2 \n",
            "  keyutils           conda-forge/linux-64::keyutils-1.6.3-hb9d3cd8_0 \n",
            "  krb5               conda-forge/linux-64::krb5-1.21.3-h659f571_0 \n",
            "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.45-default_hbd61a6d_104 \n",
            "  lerc               conda-forge/linux-64::lerc-4.0.0-h0aef613_1 \n",
            "  libblas            conda-forge/linux-64::libblas-3.11.0-4_h4a7cf45_openblas \n",
            "  libcblas           conda-forge/linux-64::libcblas-3.11.0-4_h0358290_openblas \n",
            "  libcurl            conda-forge/linux-64::libcurl-8.17.0-h4e3cde8_1 \n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.25-h17f619e_0 \n",
            "  libedit            conda-forge/linux-64::libedit-3.1.20250104-pl5321h7949ede_0 \n",
            "  libev              conda-forge/linux-64::libev-4.33-hd590300_2 \n",
            "  libexpat           conda-forge/linux-64::libexpat-2.7.3-hecca717_0 \n",
            "  libffi             conda-forge/linux-64::libffi-3.5.2-h9ec8514_0 \n",
            "  libfreetype        conda-forge/linux-64::libfreetype-2.14.1-ha770c72_0 \n",
            "  libfreetype6       conda-forge/linux-64::libfreetype6-2.14.1-h73754d4_0 \n",
            "  libgcc             conda-forge/linux-64::libgcc-15.2.0-he0feb66_16 \n",
            "  libgcc-devel_linu~ conda-forge/noarch::libgcc-devel_linux-64-15.2.0-hcc6f6b0_116 \n",
            "  libgcc-ng          conda-forge/linux-64::libgcc-ng-15.2.0-h69a702a_16 \n",
            "  libgfortran        conda-forge/linux-64::libgfortran-15.2.0-h69a702a_16 \n",
            "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-15.2.0-h69a702a_16 \n",
            "  libgfortran5       conda-forge/linux-64::libgfortran5-15.2.0-h68bc16d_16 \n",
            "  libglib            conda-forge/linux-64::libglib-2.86.3-h6548e54_0 \n",
            "  libgomp            conda-forge/linux-64::libgomp-15.2.0-he0feb66_16 \n",
            "  libiconv           conda-forge/linux-64::libiconv-1.18-h3b78370_2 \n",
            "  libjpeg-turbo      conda-forge/linux-64::libjpeg-turbo-3.1.2-hb03c661_0 \n",
            "  liblapack          conda-forge/linux-64::liblapack-3.11.0-4_h47877c9_openblas \n",
            "  liblzma            conda-forge/linux-64::liblzma-5.8.1-hb9d3cd8_2 \n",
            "  libnghttp2         conda-forge/linux-64::libnghttp2-1.67.0-had1ee68_0 \n",
            "  libopenblas        conda-forge/linux-64::libopenblas-0.3.30-pthreads_h94d23a6_4 \n",
            "  libpng             conda-forge/linux-64::libpng-1.6.53-h421ea60_0 \n",
            "  libsanitizer       conda-forge/linux-64::libsanitizer-15.2.0-h90f66d4_16 \n",
            "  libssh2            conda-forge/linux-64::libssh2-1.11.1-hcf80075_0 \n",
            "  libstdcxx          conda-forge/linux-64::libstdcxx-15.2.0-h934c35e_16 \n",
            "  libstdcxx-devel_l~ conda-forge/noarch::libstdcxx-devel_linux-64-15.2.0-hd446a21_116 \n",
            "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-15.2.0-hdf11a46_16 \n",
            "  libtiff            conda-forge/linux-64::libtiff-4.7.1-h9d88235_1 \n",
            "  libuuid            conda-forge/linux-64::libuuid-2.41.2-h5347b49_1 \n",
            "  libwebp-base       conda-forge/linux-64::libwebp-base-1.6.0-hd42ef1d_0 \n",
            "  libxcb             conda-forge/linux-64::libxcb-1.17.0-h8a09558_0 \n",
            "  libzlib            conda-forge/linux-64::libzlib-1.3.1-hb9d3cd8_2 \n",
            "  make               conda-forge/linux-64::make-4.4.1-hb9d3cd8_2 \n",
            "  ncurses            conda-forge/linux-64::ncurses-6.5-h2d0b736_3 \n",
            "  openssl            conda-forge/linux-64::openssl-3.6.0-h26f9b46_0 \n",
            "  pango              conda-forge/linux-64::pango-1.56.4-hadf4263_0 \n",
            "  pcre2              conda-forge/linux-64::pcre2-10.47-haa7fec5_0 \n",
            "  pixman             conda-forge/linux-64::pixman-0.46.4-h54a6638_1 \n",
            "  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-hb9d3cd8_1002 \n",
            "  r-base             conda-forge/linux-64::r-base-4.5.2-h835929b_2 \n",
            "  readline           conda-forge/linux-64::readline-8.2-h8c095d6_2 \n",
            "  sed                conda-forge/linux-64::sed-4.9-h6688a6e_0 \n",
            "  sysroot_linux-64   conda-forge/noarch::sysroot_linux-64-2.34-h087de78_2 \n",
            "  tk                 conda-forge/linux-64::tk-8.6.13-noxft_ha0e22de_103 \n",
            "  tktable            conda-forge/linux-64::tktable-2.10-h8d826fa_7 \n",
            "  tzdata             conda-forge/noarch::tzdata-2025b-h78e105d_0 \n",
            "  xorg-libice        conda-forge/linux-64::xorg-libice-1.1.2-hb9d3cd8_0 \n",
            "  xorg-libsm         conda-forge/linux-64::xorg-libsm-1.2.6-he73a12e_0 \n",
            "  xorg-libx11        conda-forge/linux-64::xorg-libx11-1.8.12-h4f16b4b_0 \n",
            "  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.12-hb03c661_1 \n",
            "  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.5-hb03c661_1 \n",
            "  xorg-libxext       conda-forge/linux-64::xorg-libxext-1.3.6-hb9d3cd8_0 \n",
            "  xorg-libxrender    conda-forge/linux-64::xorg-libxrender-0.9.12-hb9d3cd8_0 \n",
            "  xorg-libxt         conda-forge/linux-64::xorg-libxt-1.3.1-hb9d3cd8_0 \n",
            "  zstd               conda-forge/linux-64::zstd-1.5.7-hb78ec9c_6 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages: ...working... done\n",
            "Preparing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Verifying transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate R_env\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n",
            "Middle 95% range: 12.8302 to 24.6604 \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 25.9.1\n",
            "    latest version: 25.11.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "52HTOEzClp30"
      }
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "_b2QF-fPnmpH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda vcf_filter environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# Input VCF from 60% missingness threshold\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60.recode.vcf\"\n",
        "\n",
        "# Middle 95% Depth Percentiles (calculated earlier)\n",
        "min_depth=12.8302\n",
        "max_depth=24.6604\n",
        "\n",
        "# Output prefix\n",
        "output_prefix=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile\"\n",
        "\n",
        "# Run VCFtools to filter by mean depth\n",
        "vcftools --vcf \"$input_vcf\" \\\n",
        "         --min-meanDP \"$min_depth\" \\\n",
        "         --max-meanDP \"$max_depth\" \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix\""
      ],
      "metadata": {
        "id": "ix2BDe2tnmy-",
        "outputId": "5ed2d779-840a-443c-b7a5-f13082b7c59c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60.recode.vcf\n",
            "\t--max-meanDP 24.7\n",
            "\t--min-meanDP 12.8\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "After filtering, kept 53 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 94503 out of a possible 99691 Sites\n",
            "Run Time = 36.00 seconds\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "3ehU7lukpden"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "grep 'ZSB' vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.imiss | awk '{print $1}' > vcf_file/zsb_samples.txt"
      ],
      "metadata": {
        "id": "fMAxlS62pd4t"
      },
      "execution_count": 59,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "YmnXhyNo5PGO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda vcf_filter environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcf_filter\n",
        "\n",
        "# Define input & output\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile.recode.vcf\"\n",
        "remove_list=\"vcf_file/zsb_samples.txt\"\n",
        "\n",
        "output_prefix=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile_noZSB\"\n",
        "\n",
        "# Run vcftools to remove ZSB individuals\n",
        "vcftools --vcf \"$input_vcf\" \\\n",
        "         --remove \"$remove_list\" \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix\"\n",
        "\n",
        "echo \"Done: ZSB samples removed. Output = ${output_prefix}.recode.vcf\"\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "x0aQbZIa5hFC",
        "outputId": "a7f4c118-5f7d-43f9-c6f4-b7d3544b1916"
      },
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Done: ZSB samples removed. Output = vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile_noZSB.recode.vcf\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "VCFtools - 0.1.17\n",
            "(C) Adam Auton and Anthony Marcketta 2009\n",
            "\n",
            "Parameters as interpreted:\n",
            "\t--vcf vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile.recode.vcf\n",
            "\t--remove vcf_file/zsb_samples.txt\n",
            "\t--out vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile_noZSB\n",
            "\t--recode\n",
            "\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=DPI,Number=1,Type=Integer,Description=\"Read depth associated with indel, taken from the site preceding the indel\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=GQX,Number=1,Type=Integer,Description=\"Empirically calibrated genotype quality score for variant sites, otherwise minimum of {Genotype quality assuming variant position,Genotype quality assuming non-variant position}\">\n",
            "Warning: Expected at least 2 parts in FORMAT entry: ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Warning: Expected at least 2 parts in INFO entry: ID=BLOCKAVG_min30p3a,Number=0,Type=Flag,Description=\"Non-variant multi-site block. Non-variant blocks are defined independently for each sample. All sites in such a block are constrained to be non-variant, have the same filter value, and have sample values {GQX,DP,DPF} in range [x,y], y <= max(x+3,(x*1.3)).\">\n",
            "Excluding individuals in 'exclude' list\n",
            "After filtering, kept 48 out of 53 Individuals\n",
            "Outputting VCF file...\n",
            "After filtering, kept 94503 out of a possible 94503 Sites\n",
            "Run Time = 35.00 seconds\n"
          ]
        }
      ]
    }
  ]
}
