{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPWnqsl8CwbN09GoVkHfnZ7",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Aksinhaa/ColabFold/blob/main/NGS_collab_basic_pop_gen.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Population genetics is the study of how genetic variation is distributed within populations and how it changes over time. It connects evolution, genomics, and statistics, helping us understand how forces like mutation, migration, selection, and genetic drift shape genetic diversity.\n",
        "\n",
        "This section provides a simple, beginner-friendly overview of some of the tools used in population genetics analysis. It introduces each tool in clear, accessible language so that anyone regardless of prior experience can understand what the tools do and why they are important.\n",
        "It also prepares learners to interpret results correctly, troubleshoot problems more effectively, and apply these methods to their own datasets in the future.\n",
        "\n",
        "\n",
        "The step includes:\n",
        "\n",
        "a) Converting filtered VCF files into formats suitable for statistical analyses (e.g.,.bed, .ped) using tools like plink\n",
        "\n",
        "b) Performing Principal Component Analysis (PCA) to visualize population structure and sample clustering\n",
        "\n",
        "c) Interpreting population-level genetic structure from the PCA plots\n",
        "\n",
        "d) Running ADMIXTURE to explore population structure and ancestral relationships\n",
        "\n",
        "e) Estimating heterozygosity using RTG-tools"
      ],
      "metadata": {
        "id": "H2JOgJJjcM6W"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "### Prerequisites and Setup\n",
        "\n",
        "Before starting, make sure you have a basic understanding of SNPs, VCF files, and simple command-line usage. No local installation is required because all steps will run inside **Google Colab**.\n",
        "\n",
        "This notebook uses several population genetics tools (VCFtools, PLINK, ADMIXTURE, and RTG-tools). To keep the workflow clean and reproducible, we install everything inside a dedicated **conda environment**.\n",
        "\n",
        "When you run the setup cell, the notebook will:\n",
        "\n",
        "1. Install Miniconda\n",
        "2. Create a population genetics environment\n",
        "3. Install all required tools"
      ],
      "metadata": {
        "id": "-dk3IWPmeymT"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f6704e35"
      },
      "source": [
        "# Miniconda installation and environment setup for Colab NGS Workshop\n",
        "\n",
        "# Download and install Miniconda (skip if already installed)\n",
        "!wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh\n",
        "!bash miniconda.sh -b -p /usr/local/miniconda\n",
        "\n",
        "import sys, os\n",
        "sys.path.append('/usr/local/miniconda/lib/python3.8/site-packages')\n",
        "os.environ['PATH'] = \"/usr/local/miniconda/bin:\" + os.environ['PATH']\n",
        "\n",
        "# Accept ToS for main and R conda channels\n",
        "!conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main\n",
        "!conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r\n",
        "\n",
        "# Install necessary bioinformatics tools into the environment\n",
        "!conda create -n pop_gen -c bioconda -c conda-forge plink admixture pca rtg-tools\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cc16c8e3"
      },
      "source": [
        "First, let's list all conda environments to confirm `vcf_filter` is present."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "46f34904"
      },
      "source": [
        "!conda env list"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, we will list the packages installed in the `pop_gen` environment to ensure all th tools are there."
      ],
      "metadata": {
        "id": "NgzbBO0efKzY"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "947ce0a6"
      },
      "source": [
        "!conda list -n pop_gen"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "\n",
        "##Brief introduction to PCA(Principle Componenet Ananlysis)\n",
        "First we will perform PCA **Principal Component Analysis** which is a statistical method used to explore population structure. It reduces high-dimensional genotype data into a few key components that capture most of the genetic variation.\n",
        "\n",
        "In population genetics, PCA helps us:\n",
        "\n",
        "* Visualize genetic differences between individuals or populations\n",
        "* Detect clusters corresponding to ancestry or population groups\n",
        "* Identify outliers or mislabelled samples\n",
        "\n",
        "Essentially, PCA transforms raw genotype data into a **2D or 3D plot**, showing how individuals relate genetically.\n",
        "\n",
        "---\n",
        "\n",
        "## From VCF to PLINK Format\n",
        "\n",
        "VCF files contain SNP and genotype information but cannot be used directly for PCA in PLINK. Therefore, we need to convert VCF files into **PLINK format**, which consists of three files:\n",
        "\n",
        "| File   | Purpose                                             |\n",
        "| ------ | --------------------------------------------------- |\n",
        "| `.bed` | Binary genotype data                                |\n",
        "| `.bim` | SNP information (chromosome, position, alleles)     |\n",
        "| `.fam` | Individual/sample information (IDs, sex, phenotype) |\n",
        "\n",
        "**Workflow:**\n",
        "\n",
        "1. Start with the filtered **VCF file**\n",
        "2. Use PLINK to convert it into `.bed`, `.bim`, and `.fam` files\n",
        "3. Perform PCA on the `.bed` dataset using PLINK\n",
        "\n",
        "This step allows us to run PCA efficiently and generate plots of population structure.\n",
        "\n",
        "---\n",
        "\n",
        "# Example: Converting VCF to PLINK Format\n",
        "\n",
        "We often start population genetics analysis by converting VCF files into **PLINK binary format** (`.bed`, `.bim`, `.fam`) for downstream analyses like PCA and ADMIXTURE.\n",
        "\n",
        "Below is an example of how to run PLINK `%%bash`. This allows us to execute Bash commands directly in a Colab cell.\n",
        "\n",
        "```bash\n",
        "%%bash\n",
        "# Activate the conda environment containing PLINK\n",
        "conda activate plink\n",
        "\n",
        "# Convert a VCF file to PLINK format\n",
        "# Replace 'input_file.vcf' with your VCF filename\n",
        "# Replace 'output_prefix' with your desired output prefix\n",
        "plink --vcf input_file.vcf \\\n",
        "      --make-bed \\\n",
        "      --double-id \\\n",
        "      --allow-extra-chr \\\n",
        "      --out output_prefix\n",
        "```\n",
        "---\n",
        "\n"
      ],
      "metadata": {
        "id": "DrcEXtfWfpas"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#  Creating a Project Directory & Preparing for PCA\n",
        "\n",
        "To keep our analysis **organized and reproducible**, we will first create a separate directory to store all required files.\n",
        "\n",
        "---\n",
        "\n",
        "## Downloading Example Data\n",
        "\n",
        "For this notebook, we will use a ready-to-use dataset hosted on **Zenodo**, which contains the VCF files required for the PCA analysis.\n",
        "This ensures that everyone can **reproduce the workflow** without needing to generate raw sequencing data."
      ],
      "metadata": {
        "id": "UHryxl_emJ8u"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "c3d17aa"
      },
      "source": [
        "# Create the directory if it doesn't exist\n",
        "!mkdir -p pca\n",
        "\n",
        "# Download the VCF file into the created directory\n",
        "!wget -P pca \\\n",
        "https://zenodo.org/records/15263700/files/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB.fam\\\n",
        "https://zenodo.org/records/15263700/files/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB.recode.vcf.gz \\\n",
        "https://zenodo.org/records/15263700/files/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB.bed \\\n",
        "https://zenodo.org/records/15263700/files/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB.bim"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "At this stage, we already have the PLINK binary files (.bed, .bim, .fam) ready from our previous step. These files contain all the genotype and sample information in a format suitable for PLINK.\n",
        "\n",
        "By running PCA with PLINK, we generate two key output files:\n",
        "\n",
        "1. eigenvec → Contains the principal component scores (eigenvectors) for each individual. This file is used for plotting and visualizing genetic relationships.\n",
        "\n",
        "2. eigenval → Contains the eigenvalues corresponding to each principal component, which indicate how much genetic variation is explained by each component.\n",
        "\n",
        "In short, this step transforms our genotype data into a form that allows us to visualize population structure and relatedness across individuals."
      ],
      "metadata": {
        "id": "eNleVfWIsrc_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Initialize Conda for the current shell session\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Activate the conda environment with PLINK\n",
        "conda activate pop_gen\n",
        "\n",
        "# Perform PCA on PLINK binary files\n",
        "plink --bfile pca/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB \\\n",
        "      --pca 5 \\\n",
        "      --out pca/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB"
      ],
      "metadata": {
        "id": "9uCSgXJ4kf6m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "b4a5001a"
      },
      "source": [
        "## Install R and ggplot2 in Conda\n",
        "\n",
        "Install R-base and the `ggplot2` R package into your existing `pop_gen` conda environment. This will allow us to use R for plotting within the notebook to visualise the PCA plots.\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Initialize Conda for the current shell session\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "conda activate pop_gen\n",
        "\n",
        " # Install R-base and r-ggplot2 from specified channels\n",
        " conda install -n pop_gen -c conda-forge -c bioconda r-ggplot2 -y\n",
        "\n"
      ],
      "metadata": {
        "id": "K2JOZW_q1bS-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4c475379"
      },
      "source": [
        "**Reasoning**:\n",
        "To install R-base and the `ggplot2` R package, I need to activate the `pop_gen` conda environment and then use `conda install` to add these packages. This should be done in a single code block using `%%bash` magic command.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "hGNnZ3JSyaNW"
      },
      "source": [
        "!conda list -n pop_gen"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "939070f5"
      },
      "source": [
        "## Create R Script for PCA Plotting\n",
        "Save the provided R code for PCA plotting into a new `.R` script file.\n",
        "\n",
        "To save the provided R code into a file, we will use the `%%writefile` magic command. This will create a new file named `pca_plot.R` in the `pca/` directory with the specified R code.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0c18797a"
      },
      "source": [
        "%%writefile pca/pca_plot.R\n",
        "# Load ggplot2\n",
        "library(ggplot2)\n",
        "\n",
        "# Read FAM file from PCA directory\n",
        "fam <- read.table(\n",
        "  \"pca/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB.fam\",\n",
        "  header = FALSE\n",
        ")\n",
        "\n",
        "# Read eigenvec file from PCA directory\n",
        "eigenvec <- read.table(\n",
        "  \"pca/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_noIndels_missing_mm0.6_meandepth95percentile_noZSB.eigenvec\",\n",
        "  header = FALSE\n",
        ")\n",
        "\n",
        "# Assign column names to eigenvec\n",
        "colnames(eigenvec) <- c(\"FID\", \"IID\", paste0(\"PC\", 1:5))\n",
        "\n",
        "# Add region info from FAM (column 2)\n",
        "eigenvec$Region <- fam$V2\n",
        "\n",
        "# PCA Plot (PC1 vs PC2)\n",
        "p <- ggplot(eigenvec, aes(x = PC1, y = PC2, color = Region)) +\n",
        "  geom_point(size = 3, alpha = 0.8) +\n",
        "  theme_minimal() +\n",
        "  labs(\n",
        "    title = \"PCA Plot (PC1 vs PC2)\",\n",
        "    x = \"PC1\",\n",
        "    y = \"PC2\",\n",
        "    color = \"Region\"\n",
        "  )\n",
        "\n",
        "# Save outputs\n",
        "ggsave(\"pca/pca_plot.png\", plot = p, dpi = 300)\n",
        "ggsave(\"pca/pca_plot.pdf\", plot = p)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "b760a354"
      },
      "source": [
        "\n",
        "Now that the R script is created, we will execute it using `Rscript` within the `pop_gen` conda environment to generate the PCA plots.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eb13334f"
      },
      "source": [
        "%%bash\n",
        "# Initialize Conda for the current shell session\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Activate the conda environment with R and ggplot2\n",
        "conda activate pop_gen\n",
        "\n",
        "# Execute the R script to generate PCA plots\n",
        "Rscript pca/pca_plot.R\n"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}
