{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Aksinhaa/Pony/blob/main/NGS_collab_snp_identification.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Variant calling is the process of identifying genetic variants, such as single nucleotide polymorphisms (SNPs) and small insertions or deletions (indels) from next-generation sequencing (NGS) data. It involves comparing sequencing reads from an individual to a reference genome to detect differences, like single nucleotide variants (SNVs). In this tutorial, we will focus specifically on detecting SNPs and indels.\n",
        "\n",
        "In this step:\n",
        "\n",
        "a) Call variants (Single Nucleotide Polymorphisms and Indels) using tools like bcftools or strelka\n",
        "\n",
        "b) Produce a VCF file (Variant Call Format) with detailed information about each variant\n",
        "\n",
        "c) Optionally apply variant filtering to remove false positives or low-confidence variants\n",
        "\n",
        "Why it's important: This is the core step of population genetics—identifying the genetic differences across sample.\n"
      ],
      "metadata": {
        "id": "nDRQBoNOVaIn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "For this tutorial we are going to use Strelka, a tool utilised for germline and somatic variant calling.\n",
        "\n",
        "1: Germline Calling: Utilizes haplotype-based model to accurately detect inherited variants.\n",
        "\n",
        "(Haplotype is a set of DNA variants inherited together on the same chromosome copy)\n",
        "\n",
        "2: Somatic Calling: identifying genetic mutations that arise in somatic (non-germline) cells. These mutations are not inherited from parents and do not get passed on to offspring.\n",
        "\n",
        "Workflow Execution: Strelka2 can be run in two steps: configuration (specifying input data) and execution (specifying parameters)."
      ],
      "metadata": {
        "id": "8nEyHyy0Vn9A"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f3cc0067"
      },
      "source": [
        "\n",
        "The first step is to download the Miniconda installer for Linux using `wget`.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f6704e35",
        "collapsed": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "987378aa-a458-4dda-ea08-6f3bd9e14c2a"
      },
      "source": [
        "%%bash\n",
        "# Download and install Miniconda\n",
        "wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh\n",
        "bash miniconda.sh -b -p /usr/local/miniconda\n",
        "\n",
        "# Add conda to PATH and initialize shell\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Accept Terms of Service for required conda channels\n",
        "conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main\n",
        "conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r\n",
        "\n",
        "# Create environment and install Strelka + Samtools\n",
        "conda create -y -n strelka -c bioconda strelka samtools"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "accepted Terms of Service for https://repo.anaconda.com/pkgs/main\n",
            "accepted Terms of Service for https://repo.anaconda.com/pkgs/r\n",
            "Jupyter detected...\n",
            "2 channel Terms of Service accepted\n",
            "Channels:\n",
            " - bioconda\n",
            " - defaults\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/miniconda/envs/strelka\n",
            "\n",
            "  added / updated specs:\n",
            "    - samtools\n",
            "    - strelka\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main \n",
            "  _openmp_mutex      pkgs/main/linux-64::_openmp_mutex-5.1-1_gnu \n",
            "  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h5eee18b_6 \n",
            "  c-ares             pkgs/main/linux-64::c-ares-1.34.5-hef5626c_0 \n",
            "  ca-certificates    pkgs/main/linux-64::ca-certificates-2025.12.2-h06a4308_0 \n",
            "  certifi            pkgs/main/noarch::certifi-2020.6.20-pyhd3eb1b0_3 \n",
            "  htslib             bioconda/linux-64::htslib-1.13-h9093b5e_0 \n",
            "  krb5               pkgs/main/linux-64::krb5-1.19.2-hac12032_0 \n",
            "  libcurl            pkgs/main/linux-64::libcurl-7.87.0-h91b91d3_0 \n",
            "  libdeflate         pkgs/main/linux-64::libdeflate-1.22-h5eee18b_0 \n",
            "  libedit            pkgs/main/linux-64::libedit-3.1.20210714-h7f8727e_0 \n",
            "  libev              pkgs/main/linux-64::libev-4.33-h7f8727e_1 \n",
            "  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2 \n",
            "  libgcc             pkgs/main/linux-64::libgcc-15.2.0-h69a1729_7 \n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-15.2.0-h166f726_7 \n",
            "  libgomp            pkgs/main/linux-64::libgomp-15.2.0-h4751f2c_7 \n",
            "  libnghttp2         pkgs/main/linux-64::libnghttp2-1.52.0-ha637b67_1 \n",
            "  libssh2            pkgs/main/linux-64::libssh2-1.10.0-h37d81fd_2 \n",
            "  libstdcxx          pkgs/main/linux-64::libstdcxx-15.2.0-h39759b7_7 \n",
            "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-15.2.0-hc03a8fd_7 \n",
            "  libxcb             pkgs/main/linux-64::libxcb-1.17.0-h9b100fa_0 \n",
            "  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1 \n",
            "  openssl            pkgs/main/linux-64::openssl-1.1.1w-h7f8727e_0 \n",
            "  pip                pkgs/main/linux-64::pip-19.3.1-py27_0 \n",
            "  pthread-stubs      pkgs/main/linux-64::pthread-stubs-0.3-h0ce48e5_1 \n",
            "  python             pkgs/main/linux-64::python-2.7.18-ha1903f6_2 \n",
            "  readline           pkgs/main/linux-64::readline-8.1-h27cfd23_0 \n",
            "  samtools           bioconda/linux-64::samtools-1.13-h8c37831_0 \n",
            "  setuptools         pkgs/main/linux-64::setuptools-44.0.0-py27_0 \n",
            "  sqlite             pkgs/main/linux-64::sqlite-3.40.1-h5082296_0 \n",
            "  strelka            bioconda/noarch::strelka-2.9.10-hdfd78af_2 \n",
            "  tk                 pkgs/main/linux-64::tk-8.6.15-h54e0aa7_0 \n",
            "  wheel              pkgs/main/noarch::wheel-0.37.1-pyhd3eb1b0_0 \n",
            "  xorg-libx11        pkgs/main/linux-64::xorg-libx11-1.8.12-h9b100fa_1 \n",
            "  xorg-libxau        pkgs/main/linux-64::xorg-libxau-1.0.12-h9b100fa_0 \n",
            "  xorg-libxdmcp      pkgs/main/linux-64::xorg-libxdmcp-1.1.5-h9b100fa_0 \n",
            "  xorg-xorgproto     pkgs/main/linux-64::xorg-xorgproto-2024.1-h5eee18b_1 \n",
            "  xz                 pkgs/main/linux-64::xz-5.2.10-h5eee18b_1 \n",
            "  zlib               pkgs/main/linux-64::zlib-1.2.13-hd233ad5_2 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages: ...working... done\n",
            "Preparing transaction: | \b\b/ \b\b- \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Executing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate strelka\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR: File or directory already exists: '/usr/local/miniconda'\n",
            "If you want to update an existing installation, use the -u option.\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 25.9.1\n",
            "    latest version: 25.11.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c4c4dce4"
      },
      "source": [
        "This code block explains the setup of a Conda environment and installation of,`strelka` for variant calling.\n",
        "\n",
        "1.  **Downloading and Installing Miniconda**:\n",
        "    *   `!wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh`: This command downloads the latest Linux installer for Miniconda3. The `-q` flag ensures quiet mode (no progress bar), and `-O miniconda.sh` saves the downloaded file as `miniconda.sh`.\n",
        "    *   `!bash miniconda.sh -b -p /usr/local/miniconda`: This command executes the Miniconda installer script. The `-b` flag enables batch mode (installation process will proceed automatically without asking the user for any input), and `-p /usr/local/miniconda` specifies the installation prefix (location) as `/usr/local/miniconda`.\n",
        "\n",
        "2.  **Setting up Python Path and Environment Variables**:\n",
        "    *   `import sys, os`: Imports the `sys` and `os` modules for system-specific parameters and operating system interfaces.\n",
        "    *   `sys.path.append('/usr/local/miniconda/lib/python3.8/site-packages')`: Adds the Miniconda's Python 3.8 site-packages directory to the Python system path, allowing Python to find packages installed via Conda.\n",
        "    *   `os.environ['PATH'] = \"/usr/local/miniconda/bin:\" + os.environ['PATH']`: This ensures that Conda commands and tools installed by Conda are accessible from the command line.\n",
        "\n",
        "3.  **Installing `strelka` into the Environment**:\n",
        "    *   `!conda create -n strelka -c bioconda strelka`: This is the core installation step. It installs the `strelka`tool package into the `strelka` environment.\n",
        "        *   `-n strelka_env`: Specifies the target environment.\n",
        "        *   `-c bioconda`: Specifies `bioconda` channel as sources for packages\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        " ***Verifying `strelka` Installation***:\n",
        "      `!bash -c \"source /usr/local/miniconda/bin/activate strelka && configureStrelkaGermlineWorkflow.py\"`: This command activates the `strelka_env` environment within a new bash shell and then confirm that `strelka` is installed and functioning correctly."
      ],
      "metadata": {
        "id": "eS52l3iG6mEg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "conda list -n strelka"
      ],
      "metadata": {
        "id": "tqnV4USff2Hh",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "outputId": "ceaf6fc7-8f1b-462a-d11b-78049af66452"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# packages in environment at /usr/local/miniconda/envs/strelka:\n",
            "#\n",
            "# Name                     Version          Build            Channel\n",
            "_libgcc_mutex              0.1              main\n",
            "_openmp_mutex              5.1              1_gnu\n",
            "bzip2                      1.0.8            h5eee18b_6\n",
            "c-ares                     1.34.5           hef5626c_0\n",
            "ca-certificates            2025.12.2        h06a4308_0\n",
            "certifi                    2020.6.20        pyhd3eb1b0_3\n",
            "htslib                     1.13             h9093b5e_0       bioconda\n",
            "krb5                       1.19.2           hac12032_0\n",
            "libcurl                    7.87.0           h91b91d3_0\n",
            "libdeflate                 1.22             h5eee18b_0\n",
            "libedit                    3.1.20210714     h7f8727e_0\n",
            "libev                      4.33             h7f8727e_1\n",
            "libffi                     3.3              he6710b0_2\n",
            "libgcc                     15.2.0           h69a1729_7\n",
            "libgcc-ng                  15.2.0           h166f726_7\n",
            "libgomp                    15.2.0           h4751f2c_7\n",
            "libnghttp2                 1.52.0           ha637b67_1\n",
            "libssh2                    1.10.0           h37d81fd_2\n",
            "libstdcxx                  15.2.0           h39759b7_7\n",
            "libstdcxx-ng               15.2.0           hc03a8fd_7\n",
            "libxcb                     1.17.0           h9b100fa_0\n",
            "ncurses                    6.2              he6710b0_1\n",
            "openssl                    1.1.1w           h7f8727e_0\n",
            "pip                        19.3.1           py27_0\n",
            "pthread-stubs              0.3              h0ce48e5_1\n",
            "python                     2.7.18           ha1903f6_2\n",
            "readline                   8.1              h27cfd23_0\n",
            "samtools                   1.13             h8c37831_0       bioconda\n",
            "setuptools                 44.0.0           py27_0\n",
            "sqlite                     3.40.1           h5082296_0\n",
            "strelka                    2.9.10           hdfd78af_2       bioconda\n",
            "tk                         8.6.15           h54e0aa7_0\n",
            "wheel                      0.37.1           pyhd3eb1b0_0\n",
            "xorg-libx11                1.8.12           h9b100fa_1\n",
            "xorg-libxau                1.0.12           h9b100fa_0\n",
            "xorg-libxdmcp              1.1.5            h9b100fa_0\n",
            "xorg-xorgproto             2024.1           h5eee18b_1\n",
            "xz                         5.2.10           h5eee18b_1\n",
            "zlib                       1.2.13           hd233ad5_2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "For the variant calling step, we will download the BAM file which is aligned, sorted, and duplicate-marked file (generated during the mapping step), from the Zenodo repository. This will ensure that they are ready for use in downstream variant identification.\n"
      ],
      "metadata": {
        "id": "41r7Kk_gDRvu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka\n",
        "\n",
        "which configureStrelkaGermlineWorkflow.py"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uALeX4fFDiYd",
        "outputId": "59899115-c820-4684-c454-e55fb3ed8825"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/usr/local/miniconda/envs/strelka/bin/configureStrelkaGermlineWorkflow.py\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "rm -rf reference\n",
        "\n",
        "# URL for reference genome tar.gz file\n",
        "reference_tar=\"https://zenodo.org/records/17878528/files/reference.tar.gz?download=1\"\n",
        "\n",
        "# Download the tar.gz file\n",
        "wget -q \"$reference_tar\" -O reference.tar.gz\n",
        "\n",
        "# Extract the reference folder\n",
        "tar -xzvf reference.tar.gz\n",
        "\n",
        "# List extracted contents\n",
        "ls -F reference/"
      ],
      "metadata": {
        "id": "t72HE-onDRIf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "45fe446b-406d-43d1-fce4-67db0426b025"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "reference/\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.ann\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.amb\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.sa\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.bwt\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.pac\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "reference/.ipynb_checkpoints/\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.amb\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.ann\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.bwt\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.pac\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.sa\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Once we have our required BAM files, we will start the step for variant calling. In this analysis we use the Strelka germline workflow, which is initiated through the configureStrelkaGermlineWorkflow.py script.\n",
        "\n",
        "This germline configuration file contains:\n",
        "\n",
        "1: Information about the input BAM files (aligned, sorted, and duplicate-marked reads).\n",
        "\n",
        "2: The reference genome path, which Strelka uses for aligning sequences and identifying variant sites.\n",
        "\n",
        "3: Workflow parameters, such as filtering settings, runtime options, and rules for how Strelka processes sequencing data.\n",
        "\n",
        "4: Module definitions, specifying the order and structure of the processing steps used to detect SNVs and small indels.\n",
        "\n",
        "5: Paths to output directories, where Strelka will write variant call results, logs, and intermediate files.\n",
        "\n",
        "In short, this germline configuration is used by Strelka to build and execute the variant calling workflow. It ensures that the pipeline runs consistently with the correct inputs, reference genome, and settings needed for accurate germline variant detection."
      ],
      "metadata": {
        "id": "kOdeQQ8NGRC0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Create bam_files directory if it doesn't exist\n",
        "mkdir -p bam_files\n",
        "\n",
        "# Move the .fna file from reference/ to bam_files/\n",
        "mv reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna bam_files/\n",
        "\n",
        "# List contents of bam_files to confirm\n",
        "ls -F bam_files"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "RY3TepNr9WYE",
        "outputId": "06d8a2c1-bdf1-469f-caf7-7d97ce370ae2"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "BEN_CI16.bam\n",
            "BEN_CI16.bam.bai\n",
            "BEN_NW10.bam\n",
            "BEN_NW10.bam.bai\n",
            "BEN_SI18.bam\n",
            "BEN_SI18.bam.bai\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "9SUB-qF4-pGF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda environment that contains samtools\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka   # or the environment where samtools is installed\n",
        "\n",
        "# Move into bam_files directory\n",
        "cd bam_files\n",
        "\n",
        "# Index the reference FASTA\n",
        "samtools faidx GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
        "\n",
        "echo \"Indexed reference file: $fna_file\"\n",
        "ls -F\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0hO1HqtS_V9p",
        "outputId": "24a08cbc-4492-4373-eaeb-2b6608419228"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Indexed reference file: \n",
            "BEN_CI16.bam\n",
            "BEN_CI16.bam.bai\n",
            "BEN_NW10.bam\n",
            "BEN_NW10.bam.bai\n",
            "BEN_SI18.bam\n",
            "BEN_SI18.bam.bai\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.fai\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "au3HYKKI_WMP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Create directory for BAM files\n",
        "mkdir -p bam_files\n",
        "\n",
        "# Download all BAM and BAI files into bam_files/\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_CI16_aligned_reads_deduplicated.bam?download=1\"     -O bam_files/BEN_CI16.bam\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_CI16_aligned_reads_deduplicated.bam.bai?download=1\" -O bam_files/BEN_CI16.bam.bai\n",
        "\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_NW10_aligned_reads_deduplicated.bam?download=1\"     -O bam_files/BEN_NW10.bam\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_NW10_aligned_reads_deduplicated.bam.bai?download=1\" -O bam_files/BEN_NW10.bam.bai\n",
        "\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_SI18_aligned_reads_deduplicated.bam?download=1\"     -O bam_files/BEN_SI18.bam\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_SI18_aligned_reads_deduplicated.bam.bai?download=1\" -O bam_files/BEN_SI18.bam.bai\n",
        "\n",
        "# List downloaded files\n",
        "ls -lh bam_files\n"
      ],
      "metadata": {
        "id": "ejHSeGlmLNvu",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0daba8b1-6806-4fb8-ae9e-676b14d1bdf4"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 843M\n",
            "-rw-r--r-- 1 root root 272M Dec 11 15:46 BEN_CI16.bam\n",
            "-rw-r--r-- 1 root root 3.3M Dec 11 15:47 BEN_CI16.bam.bai\n",
            "-rw-r--r-- 1 root root 259M Dec 11 15:49 BEN_NW10.bam\n",
            "-rw-r--r-- 1 root root 3.3M Dec 11 15:49 BEN_NW10.bam.bai\n",
            "-rw-r--r-- 1 root root 303M Dec 11 15:54 BEN_SI18.bam\n",
            "-rw-r--r-- 1 root root 3.1M Dec 11 15:54 BEN_SI18.bam.bai\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "\n",
        "After running the Strelka germline variant calling workflow, the pipeline produces a set of output files, most importantly the **VCF (Variant Call Format) files**. These files contain the list of single nucleotide variants (SNVs) and small insertions/deletions (indels) identified from the input BAM files.\n",
        "\n",
        "Once the VCF files are generated, they represent the final variant calls produced by Strelka, and they can now be used for downstream analyses. Typical post-processing steps may include:\n",
        "\n",
        "* **Quality assessment** of the variants by inspecting Strelka’s FILTER and quality score annotations.\n",
        "* **Merging or comparing VCFs** (if multiple samples were processed) to evaluate shared or unique variants.\n",
        "* **Functional annotation** using external tools (e.g., VEP, SnpEff) to predict gene impacts, consequences, or biological relevance.\n",
        "* **Filtering based on depth, genotype quality, or allele frequency**, depending on the research goals.\n",
        "\n",
        "In summary, the generation of VCF files marks the completion of the variant calling workflow, providing a structured dataset of all detected variants. These files form the foundation for interpretation, annotation, and further biological or population-level analyses.\n"
      ],
      "metadata": {
        "id": "XyaYBoxiLNH3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka\n",
        "\n",
        "# Create the run directory manually (recommended for clarity)\n",
        "mkdir -p strelka_run\n",
        "\n",
        "# Configure Strelka workflow\n",
        "configureStrelkaGermlineWorkflow.py \\\n",
        "    --bam bam_files/BEN_CI16.bam \\\n",
        "    --bam bam_files/BEN_NW10.bam \\\n",
        "    --bam bam_files/BEN_SI18.bam \\\n",
        "    --referenceFasta bam_files/GCA_021130815.1_PanTigT.MC.v3_genomic.fna \\\n",
        "    --runDir strelka_run"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "vsYRDIa59Bj3",
        "outputId": "28f71c12-9293-47d1-ad1b-ff78e2d612b9"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Successfully created workflow run script.\n",
            "To execute the workflow, run the following script and set appropriate options:\n",
            "\n",
            "/content/strelka_run/runWorkflow.py\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka\n",
        "\n",
        "strelka_run/runWorkflow.py -m local -j 8"
      ],
      "metadata": {
        "id": "HQwRwuGNF-hz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "cJ8qzZir9B7r"
      }
    }
  ]
}
