{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Aksinhaa/ColabFold/blob/main/NGS_collab_Linux_advanced.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "88b3a0d7"
      },
      "source": [
        "## Introduction\n",
        "\n",
        "This notebook provides a comprehensive guide and analysis of sequencing datasets. We will cover various tasks, including data download, understanding reference index files, exploring the structure of FASTQ files, and performing several quality control and analysis steps on sequencing reads. Each section is designed to walk you through the process, from basic file operations to more advanced data interrogation.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l0jsQPYkNf2V"
      },
      "source": [
        "# Sequencing Dataset Instructions"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6sqdVIKHO8HC"
      },
      "source": [
        "**Task 1** : Downloads\\\n",
        "Download the datasets from https://zenodo.org/records/14258052 .\\\n",
        "Fastq reads are in `.gz `format. Original genome files are in `.fna` (FASTA) format."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "743d02e0"
      },
      "source": [
        "### Downloading the FASTQ files\n",
        "\n",
        "I will download the specific FASTQ files mentioned in the previous error messages directly from the Zenodo repository. Please note that the Zenodo URL structure for specific files is often `https://zenodo.org/record/{record_id}/files/{filename}`."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zhn5rl8NRTTr"
      },
      "source": [
        "**Task 2** : Reference Index Sidecar Files\n",
        "\n",
        "*  ` .fna.amb`\n",
        ": stores information about ambiguous bases (e.g., N) in the reference genome.\n",
        "\n",
        "*   `.fna.ann`: stores information about sequence names, lengths, and other metadata.\n",
        "*   `.fna.bwt`: contains the Burrowsâ€“Wheeler transformed sequence.\n",
        "\n",
        "\n",
        "\n",
        "*  `.fna.pac`: contains packed sequence data.\n",
        "\n",
        "*   `.fna.sa`: suffix array index, used to locate sequence positions.\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OS7O7K5eTfZ7"
      },
      "source": [
        "**Task 3**: Structure of a FASTQ File\n",
        "\n",
        "FASTQ files are obtained after demultiplexing Illumina sequencing data. Providers usually deliver them as `.fastq.gz` for compression. \\\n",
        "Each record has four lines:\n",
        "\n",
        "\n",
        "```\n",
        "# @SRR15369215.126490887  # sequence identifier\n",
        "GGACCTTCTGTCATTTCACTCCTTCTGAAGTAAGGAGTGAAGTAAACACGAAGTAAACACGACAGGTTAGTCCTATTCCTTCAAGCAGGAGTACAGAAAAGAATGCAAATTCTGGGTTCTAGCCCAGCTTTTACTCCTATGGTTCTATTT  # sequence\n",
        "+  #separator\n",
        "AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJFJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFAFJJJJJJJJJJJJJJJJFJJJJJJJJJFJJJJJJJJJFFFJJJJJJJJJJJ  # base call quality scores (ASCII characters)\n",
        "```\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Download the files\n",
        "\n",
        "%%bash\n",
        "wget https://zenodo.org/records/14258052/files/BEN_CI16_sub_1.fq.gz\n",
        "wget https://zenodo.org/records/14258052/files/BEN_CI16_sub_2.fq.gz\n",
        "\n"
      ],
      "metadata": {
        "collapsed": true,
        "id": "M-_6EWtcOqY9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#For decompressing all the downloaded files\n",
        "\n",
        "%%bash\n",
        "\n",
        "gunzip *fq.gz"
      ],
      "metadata": {
        "id": "q9r0IRjhTLEa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "View the first 20 lines of a fastq file. How many seqeunces are present? What does each line represent"
      ],
      "metadata": {
        "id": "M9C2dq4dEAPj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "head -n 20 BEN_CI16_sub_1.fq"
      ],
      "metadata": {
        "id": "39o5jOkpEKtm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VMRaGVXCVLHI"
      },
      "source": [
        "#Question 1\n",
        "\n",
        "Count the number of sequencing reads in a FASTQ files"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "q5hZbNLr4Kbr"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "less BEN_CI16_sub_1.fq | awk 'NR % 4 == 1' | wc -l\n",
        "#less: views the files.\n",
        "#awk 'NR % 4 == 1': Filters for lines where the record number (NR) modulo 4 is 1, which corresponds to the header lines (starting with @).\n",
        "#wc -l: Counts the number of lines, effectively counting the number of reads.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Count the number of reads in all FASTQ files"
      ],
      "metadata": {
        "id": "l8KWHm_SHKUk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "#bulk analysis\n",
        "for file in *.fq; do\n",
        "  read_count=$(( $(less \"$file\" | wc -l) / 4 ))\n",
        "  echo \"$(basename \"$file\"): $read_count reads\"\n",
        "done"
      ],
      "metadata": {
        "id": "hOfAg2ooVk6V"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HBEidGJcVZAn"
      },
      "source": [
        "#Question 2\n",
        "\n",
        "How many reads are shorter than 150 bp?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BslH8S9uAa0Y"
      },
      "outputs": [],
      "source": [
        "#Question 2 solution\n",
        "\n",
        "%%bash\n",
        "less BEN_CI16_sub_1.fq | awk 'NR % 4 == 2 {print length($0)}' | awk '$1<150' | wc -l\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### How many reads are shorter than 150 bp in all the FASTQ files:"
      ],
      "metadata": {
        "id": "sSfxAwDBHgiE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "#bulk analysis\n",
        "for file in *.fq; do\n",
        "   reads=$(less \"$file\" | awk 'NR % 4 == 2 {print length($0)}' | awk '$1<150' | wc -l)\n",
        "   echo \"$(basename \"$file\"): $reads reads shorter than 150bp\"\n",
        " done\n"
      ],
      "metadata": {
        "id": "uZP56gNNVhMY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Mm_vp0sdVjb-"
      },
      "source": [
        "#Question 3\n",
        "\n",
        "How many reads are longer than 150 bp?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lEOwCXyFB-n7"
      },
      "outputs": [],
      "source": [
        "#Question 3 solution\n",
        "\n",
        "%%bash\n",
        "less BEN_CI16_sub_1.fq | awk 'NR % 4 == 2 {print length($0)}' | awk '$1>150' | wc -l\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### How many reads are longer than 150 bp in all the FASTQ files:"
      ],
      "metadata": {
        "id": "qChbXe6HHw3p"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "#bulk analysis\n",
        "for file in *.fq; do\n",
        "   reads=$(less \"$file\" | awk 'NR % 4 == 2 {print length($0)}' | awk '$1>150' | wc -l)\n",
        "   echo \"$(basename \"$file\"): $reads reads longer than 150bp\"\n",
        " done"
      ],
      "metadata": {
        "id": "HW_yLBUfVdnF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "O08qtzp7VqB1"
      },
      "source": [
        "#Question 4\n",
        "\n",
        "How many reads are not equal to 150 bp?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "1hksra_MCO0c"
      },
      "outputs": [],
      "source": [
        "#Question 4 solution\n",
        "\n",
        "%%bash\n",
        "less BEN_CI16_sub_1.fq | awk 'NR % 4 == 2 {if (length($0) != 150) print length($0)}' | wc -l\n",
        "#if (length($0) != 150): Checks if the length of the sequence is not equal to 150.\n",
        "#print length($0): Prints the length of sequences that are not 150.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### How many reads are equal to 150 bp in all the FASTQ files:"
      ],
      "metadata": {
        "id": "jmu109jyH6pd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "#bulk analysis\n",
        "for file in *.fq; do\n",
        "  count=$(less \"$file\" | awk 'NR % 4 == 2 {if (length($0) != 150) print length($0)}' | wc -l)\n",
        "  echo \"$(basename \"$file\"): $count reads not equal to 150bp\"\n",
        "done"
      ],
      "metadata": {
        "id": "cE8Tt26GVYKg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YU5uKAiqVq4m"
      },
      "source": [
        "#Question 5\n",
        "\n",
        "How many reads are contaminated with Illumina adapters?\n",
        "\n",
        "\n",
        "\n",
        "```\n",
        "# Adapter sequence: CTGTCTCTTATACACATCT\n",
        "```\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8_5QH27PCXBq"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "count=$(less BEN_CI16_sub_1.fq | awk 'NR % 4 == 2' | grep -c 'CTGTCTCTTATACACATCT')\n",
        "echo \"$count\""
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###To check how many reads are contaminated with Illumina adapters in all the FASTQ files:"
      ],
      "metadata": {
        "id": "o7TjPUrgIEMz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "#bulk analysis\n",
        "for file in *.fq; do\n",
        "    count=$(less \"$file\" | awk 'NR % 4 == 2' | grep -c 'CTGTCTCTTATACACATCT')\n",
        "    echo \"$file: $count\"\n",
        "done"
      ],
      "metadata": {
        "id": "WWZNoH51VOvk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Tutorial Questions**\n",
        "1) Download the following files and count the numbers of reads more than, less than and equal to 100bp in the given empty code cell.\n",
        "\n",
        "https://zenodo.org/records/14258052/files/BEN_NW10_sub_1.fq.gz\n",
        "\n",
        "https://zenodo.org/records/14258052/files/BEN_NW10_sub_2.fq.gz\n",
        "\n",
        "\n",
        "https://zenodo.org/records/14258052/files/BEN_SI18_sub_1.fq.gz\n",
        "\n",
        "https://zenodo.org/records/14258052/files/BEN_SI18_sub_2.fq.gz\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "Ymlxds72l-n5"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Kz_eWNTeErI8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Answer the following questions in the below text cell:\n",
        "\n",
        "1) Why does counting reads in a FASTQ file involve dividing the total number of lines by 4?\n",
        "2) How does sequencing quality typically vary along the length of a read, and why?\n",
        "3) What biological or technical factors contribute to low-quality base calls?\n",
        "4) Conceptually, how does insert size constrain correct mapping locations?\n",
        "5) If a dataset has many bases with Phred scores below Q20, how would this affect downstream mapping?\n"
      ],
      "metadata": {
        "id": "ZR8QtaO6EryJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "s7xhJZbJFAst"
      }
    }
  ]
}
